---
title: "2020_Metabarcoding_Analyses"
author: "Georgia Titcomb"
date: "January 31, 2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(vegan)
library(GUniFrac)
library(sjPlot)
library(readxl)
library(ggConvexHull)
library(igraph)
library(phylosignal)
library(adephylo)
library(ape)
library(phylobase)
library(philentropy)
library(ComplexHeatmap)
library(lmerTest)
library(emmeans)


```

## 1. Import data and finish filtering

### 1.1 Setup and visualization
```{r}


OTUtab <- read.csv("allMOTUs_sumatra99_GTmethod2020.csv")
dim(OTUtab)

# sequencing depth
median(rowSums(OTUtab[,-c(1,1703:1704)])) 
mean(rowSums(OTUtab[,-c(1,1703:1704)]))
hist(rowSums(OTUtab[,-c(1,1703:1704)]))

# Remove samples with reads = 0
OTUtabsmall <- OTUtab[-which(rowSums(OTUtab[,-c(1,1703:1704)])==0),]
dim(OTUtabsmall)

# calculate relative reads per sample
RRAtab <- OTUtabsmall %>% mutate_at(vars(MOTU_1:MOTU_1701), funs(./total_reads))

# mutate rows that have RRA less than 1% to 0
RRAtab_filt <- RRAtab %>% mutate_at(vars(MOTU_1:MOTU_1701), funs(ifelse(.<0.01,0,.)))

# How many OTUs are now zero?
length(which(colSums(RRAtab_filt[,2:1702])==0))

# Remove those columns
RRAtab_filt2 <- RRAtab_filt[,-(which(colSums(RRAtab_filt[,2:1702])==0)+1)]
dim(RRAtab_filt2)


mean(RRAtab_filt2$total_reads)
median(RRAtab_filt2$total_reads)
min(RRAtab_filt2$total_reads)

hist(RRAtab_filt2$total_reads)
ggplot(RRAtab_filt2, aes(x=total_reads))+
  geom_histogram(binwidth=0.2)+
  scale_x_log10(breaks=c(1,10,100,1000,10000))+
  geom_vline(xintercept=100, col="red", size=2)+
  theme_bw(base_size=14)

# 100 seems to be a reasonable threshold (this is after adjusting for negative controls)


```

### 1.2 Filtering

```{r}


# Remove samples with reads<100. I used the Albatross paper for reference here
OTUtabsmall <- OTUtab[-which(rowSums(OTUtab[,-c(1,1703:1704)])<100),]

dim(OTUtabsmall)
#names(OTUtabsmall)

# Recalculate RRA
RRAtab <- OTUtabsmall %>% mutate_at(vars(MOTU_1:MOTU_1701), funs(./total_reads))

# Filter out >1% RRA
RRAtab_filt <- RRAtab %>% mutate_at(vars(MOTU_1:MOTU_1701), funs(ifelse(.<0.01,0,.)))

# How many OTUs are now zero?
length(which(colSums(RRAtab_filt[,2:1702])==0))

# Remove those columns
RRAtab_filt2 <- RRAtab_filt[,-(which(colSums(RRAtab_filt[,2:1702])==0)+1)]
dim(RRAtab_filt2)

# Currently we have 487 MOTUs in 676 samples
mean(RRAtab_filt2$total_reads)
median(RRAtab_filt2$total_reads)
min(RRAtab_filt2$total_reads)

ggplot(RRAtab_filt2, aes(x=total_reads))+
  geom_histogram(binwidth=0.1)+
  scale_x_log10(breaks=c(1,10,100,1000,10000))+
  geom_vline(xintercept=100, col="red", size=2)+
  theme_bw(base_size=14)

```

### 1.3 Relationship between reads and richness -- is rarefaction warranted?

```{r}
RRAtab_filt2$richness <- specnumber(RRAtab_filt2[,2:488])

ggplot(RRAtab_filt2, aes(x=log(RRAtab_filt2$total_reads), y=richness))+
  geom_jitter(width=0.5,height=0.5)+
  geom_smooth(method="lm")

```


### 1.4 Rarefying the data

From vegan:  Random rarefaction is sometimes used to remove the effects of different sample sizes. This is usually a bad idea: random rarefaction discards valid data, introduces random error and reduces the quality of the data (McMurdie & Holmes 2014). It is better to use normalizing transformations (decostand in vegan) possible with variance stabilization (decostand and dispweight in vegan) and methods that are not sensitive to sample sizes.

```{r}

#OTUrare=decostand(OTUtabsmall[,2:569], 'total')
#OTUrare$total_reads = OTUtabsmall$total_reads
#OTUtabsmall$rownames = row.names(OTUtabsmall)
#OTUrare$rownames = row.names(OTUrare)

# Combine information into one df
#OTUrare = left_join(OTUtabsmall[,c(572,1)],OTUrare)

# Recalculate RRA (not needed with decostand function)
#RRAtab <- OTUrare %>% mutate_at(vars(MOTU_1:MOTU_568), funs(./total_reads))

# Filter
# RRAtab_filt <- RRAtab_filt2 %>% mutate_at(vars(MOTU_1:MOTU_557), funs(ifelse(.<0.01,0,.)))
# dim(RRAtab_filt)
# 175 OTUs, 663 samples # note that there is variation in the OTU

mean(RRAtab_filt2$total_reads, na.rm=T)
median(RRAtab_filt2$total_reads, na.rm=T)
min(RRAtab_filt2$total_reads, na.rm=T)

names(RRAtab_filt2)[1]="Sample_ID"

# save a copy
#write.csv(RRAtab_filt2, "RRAtab_filt2_99.csv", row.names=F)

```


## 2. Initial richness comparison

### 2.1 Read in and join datasets

```{r}

RRA_rare <- read.csv("RRAtab_filt2_99.csv")
#head(RRA_rare)
RRA_hosts <- read_excel("SAMPLE_SUMMARY.xlsx",sheet="Sheet1")

# Combine MOTU table with host data
RRA2 = full_join(RRA_hosts, RRAtab_filt2)
dim(RRA2)

RRA2[which(is.na(RRA2$Location)),] # only one duplicate sample has no associated info
RRA2=RRA2[-which(is.na(RRA2$Location)),]

# Use only the qPCR prevalence
RRA2 = RRA2[which(RRA2$rd_2_qpcr_conducted=="Y"),]
dim(RRA2)

# remove animals not focused on
RRA2 = RRA2[-which(RRA2$Species %in% c("Bushbuck","NTC","Goat","Sheep","EC","Oribi")),]
# All NA values indicate additional zeros

names(RRA2)
# select columns of interest
RRA2 = RRA2[,c(1:14,16:502)]
head(RRA2)

# Replace NA values with 0 for samples that were undetected
# This includes things that were not screened
RRA2 = RRA2 %>% mutate_at(vars(MOTU_1:MOTU_1677), funs(ifelse(is.na(.),0,.)))
names(RRA2)
```

### 2.2 Calculate richness

```{r}

# Add MOTU richness for each sample (after rarifying to same reads)
RRA2$richness = specnumber(RRA2[,15:501])

# Calculate species level richness
sumR = RRA2 %>% group_by(Location, Species) %>% summarize_at(vars(MOTU_1:MOTU_1677),funs(sum(.,na.rm=T)))
richy = specnumber(sumR[,3:489])

# Determine the number of samples examined
numR = RRA2 %>% group_by(Location, Species) %>% summarize(count=n())
numR$total_richness=richy

# Determine the average individual richness by species
indR = RRA2 %>% group_by(Location, Species) %>% summarize_at(vars(richness), funs(mean(.,na.rm=T)))

# add to df
numR$ind_rich = indR$richness
head(numR)

# population richness related to sample count?
ggplot(numR, aes(x=count, y=total_richness))+
  geom_point(aes(color=Location))+
  facet_wrap(~Species)+
  theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0))
# yes

ggplot(numR, aes(x=count, y=ind_rich))+
  geom_point(aes(color=Location))+
  facet_wrap(~Species)


# Relationship between mean individual and total richness
ggplot(numR, aes(x=ind_rich, y=total_richness))+
  geom_point(aes(color=Species, size=count, shape=Location))+
  geom_hline(yintercept=mean(numR$total_richness))+
  geom_vline(xintercept=mean(numR$ind_rich))+
  scale_shape_manual(values=11:19)+
  annotate(geom="text",x=1.5,y=20,label="Relatively over sampled")+
  annotate(geom="text",x=6,y=4,label="Relatively under sampled")

```


```{r}
# define palette for all species

animal_colors=c("#6CA6CD", "#CD919E", "#EE799F", "#A2CD5A", "#7EC0EE", "#CD6090", "#63B8FF", "#9BCD9B", "#CD8C95", "#4F94CD", "#8DB6CD", "#6495ED", "#CDC673", "#BDB76B", "#7AC5CD", "#9370DB", "#6A5ACD", "#7CCD7C", "#87CEFF")

RRA2$Species = as.factor(RRA2$Species)
RRA2$Species = factor(RRA2$Species, levels=levels(reorder(RRA2$Species,-RRA2$richness)))

withzeros = ggplot(RRA2, aes(x=reorder(Species,-richness), y=richness))+
  geom_jitter(width=0.01,height=0.3, alpha=0.5, aes(color=Species))+
  geom_boxplot(aes(fill=Species), color="gray50",outlier.shape=NA)+
  guides(fill=F,col=F)+
  theme_bw()+
  scale_fill_manual(values=animal_colors)+
  scale_color_manual(values=animal_colors)+
  theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0))+
  labs(x="Host Species",y="Rarified OTU richness",title="Individual-level Richness (with 0s)")

nozeros = ggplot(RRA2[RRA2$richness!=0,], aes(x=reorder(Species,-richness,na.rm=T), y=richness))+
  geom_jitter(width=0.01,height=0.3, alpha=0.5, aes(color=Species))+
  geom_boxplot(aes(fill=Species), color="gray50",outlier.shape=NA)+
  guides(fill=F,col=F)+
  theme_bw()+
  scale_fill_manual(values=animal_colors)+
  scale_color_manual(values=animal_colors)+
  theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0))+
  labs(x="Host Species",y="Rarified OTU richness",title="Individual-level Richness (no 0s)")

gridExtra::grid.arrange(withzeros, nozeros, ncol=2)
```

The plot above shows the rarified OTU richness for two scenarios: 1) Including zeros for any sample that was tested and did not 'make it' to the filtering step. 2) Excluding all zeros.


### Prevalence diagram

```{r}
head(RRA2)

counts = RRA2 %>% group_by(Species, Location) %>% summarize(N = n())
pressies = RRA2 %>% group_by(Species, Location) %>% summarize_at( vars(richness),funs(count(ifelse(.>0,1,0))))
prevdf = merge(counts,pressies)
prevdf$prevalence=prevdf$richness/prevdf$N

head(colorsp)
prevdf = left_join(prevdf,colorsp, by=c("Species"="Host"))

# prevdf[(prevdf$prevalence)==0,]$prevalence = c(0.001, 0.002,0.003,0.004,0.005,0.006)
# prevdf[(prevdf$prevalence)==0.5,]$prevalence =c(0.5,0.5001)
# prevdf[(prevdf$prevalence)==0.1,]$prevalence = c(0.1,0.1001)
# prevdf[(prevdf$prevalence)==0.2,]$prevalence = c(0.2,0.2001)
# prevdf[(prevdf$prevalence)==0.125,]$prevalence = c(0.125,0.1251)
# prevdf[(prevdf$prevalence)==0.25,]$prevalence = c(0.25,0.2501)
# prevdf[(prevdf$prevalence)==9/21,]$prevalence = c(0.4285714,0.4285715)
# prevdf[(prevdf$prevalence)==8/22,]$prevalence = c(0.3636364,0.3636365)
# prevdf[(prevdf$prevalence)==1,]$prevalence = c(0.9999998, 0.99999999,1)
# prevdf[(prevdf$prevalence)==7/19,]$prevalence = c(0.3684211,0.3684212)


#prevdf$colors = factor(prevdf$colors, levels=reorder(prevdf$colors,-prevdf$prevalence))
prevdf = prevdf[order(prevdf$prevalence),]
str(prevdf)
levels(as.factor(prevdf$Species))
levels(prevdf$colors)
levels(as.factor(prevdf$Species))

levels(as.factor(prevdf$Species))

newcols = data.frame(species = levels(RRA_sp$Species), colors=animal_colors)
newcols$species = factor(newcols$species, levels=levels(RRA_sp$Species))

ggplot(newcols, aes(x=species, y=1))+
  geom_col(aes(fill=species))+
  scale_fill_manual(values=as.character(newcols$colors))

prevdf$Species = factor(prevdf$Species, levels=levels(RRA_sp$Species))
prevdf$Location = factor(prevdf$Location, levels=c("Mpala","Kruger","Hwange","Kafue","Hluhluwe","Serengeti","Nyika","Niassa","Gorongosa"))

ggplot(prevdf[-which(prevdf$Species=="Hippo"&prevdf$Location=="Kafue"),], aes(x=reorder(Species,-prevalence), y=prevalence))+
  geom_col(aes(fill=Species), position="dodge")+
  geom_text(aes(label=richness),nudge_y=0.07)+
  facet_wrap(~Location,scales="free_x")+
  theme_bw(base_size=14)+
  guides(fill=F)+
  scale_fill_manual(values=as.character(newcols$colors))+
  theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0))+
  labs(x="Host Species",y="Parasite Prevalence (any OTU)")

```


### 2.3 Conduct statistical tests of richness

```{r}

# Does individual richness vary significantly by species? -- Question for MP data
# not including zero data:
RRA_present = RRA2[which(RRA2$richness>0),]
MPdata = RRA_present[RRA_present$Location=="Mpala",]

# show distribution
hist(MPdata$richness)

########################
# fit lmer, prob not going to be great
#richmodMP = lmer(richness ~ Species + (1|Period), data=MPdata)
#summary(richmodMP)
#plot_model(richmodMP, "diag") # non normal

# use poisson distribution
#richmodMP = glmer(richness ~ Species + (1|Period),family="poisson", data=MPdata)
#summary(richmodMP) # deviance > df residuals, still overdispersed from 0s probably

# use nb distribution
richmodMP = glmer.nb(richness ~ Species+(1|Period), data=MPdata, trace=F)
summary(richmodMP) # deviance >> df residuals, still overdispersed
car::Anova(richmodMP)

library(DHARMa)
testResiduals(simulateResiduals(richmodMP)) ### this is the best model

tab_model(richmodMP,transform=NULL, show.se=T, show.stat=T, show.r2=T)

##################################
#### Including 0s

MPdata_0 = RRA2[which(RRA2$Location=="Mpala"),]

# show distribution
hist(MPdata_0$richness)

# fit lmer, prob not going to be great
#richmodMP0 = lmer(richness ~ Species + (1|Period), data=MPdata_0)
#summary(richmodMP0)
#plot_model(richmodMP0, "diag") # non normal

# use poisson distribution
#richmodMP0 = glmer(richness ~ Species + (1|Period),family="poisson", data=MPdata_0)
#summary(richmodMP0) # deviance > df residuals, still overdispersed from 0s probably

# use nb distribution
#richmodMP0 = glmer.nb(richness ~ Species + (1|Period), data=MPdata_0)
#summary(richmodMP0) # deviance >> df residuals, still overdispersed

#testResiduals(simulateResiduals(richmodMP0))
#MuMIn::r.squaredGLMM(richmodMP0)

# use zero-inflated method
library(glmmTMB)

# little zero inflation variation when modeled by species, so changed to 1
richmodMP0.1 = glmmTMB(richness ~ Species+(1|Period), ziformula=~Species, family="nbinom2",data=MPdata_0,REML=T)

richmodMP0 = glmmTMB(richness ~ Species+(1|Period), ziformula=~1, data=MPdata_0, family="nbinom2", REML=T)

bbmle::AICctab(richmodMP0.1,richmodMP0)
# convergence problems with REML=F, comparable zi results, so z formula is 1

#MuMIn::r.squaredGLMM(richmodMP0)

testResiduals(simulateResiduals(richmodMP0))

summary(richmodMP0)
car::Anova(richmodMP0, test.statistic="Chisq")


tab_model(transform=NULL,richmodMP0, show.se=T, show.stat=T, show.r2=T,show.p=T)



```

### 2.4 Perform post-hoc comparison

```{r}
# no zeros
ems = emmeans(richmodMP, ~Species)
CLD(ems)
contrast(ems, adjust="holm")

MuMIn::r.squaredGLMM(richmodMP)
# Nearly 60% of variation in richness explained by host species for Mpala dataset

# with zeros
ems0 = emmeans(richmodMP0, ~Species)
CLD(ems0)
contrast(ems0, adjust="holm")
#MuMIn::r.squaredGLMM(richmodMP0) # not available with zero inflation

```


### Richness by other host attributes

```{r}

host_info = read.csv("host_attributes_2020.csv")
head(host_info)

MP2 = MPdata_0 %>% group_by(Species) %>% summarize_at(vars(richness), funs(mean))
MP2 = left_join(MP2,host_info)

ggplot(MP2, aes(x=log(Body_Mass_kg), y=richness))+
  geom_point()+
  geom_smooth(method="lm")

ggplot(MP2, aes(x=log(SocialGrpSize), y=richness))+
  geom_point()+
  geom_smooth(method="lm")

ggplot(MP2, aes(x=log(HomeRange_km2), y=richness))+
  geom_point()+
  geom_smooth(method="lm")

ggplot(MP2, aes(x=Diet, y=richness))+
  geom_boxplot()

richmoddy = lm(richness ~ Diet*log(HomeRange_km2)+log(SocialGrpSize)+log(Body_Mass_kg)+MSW93_Family, data=MP2)

car::Anova(richmoddy)
summary(richmoddy)

richmoddy = lm(richness ~ Diet+log(HomeRange_km2)+log(SocialGrpSize)+MSW93_Family, data=MP2)
car::Anova(richmoddy,type=3)


richmoddy2 = lm(richness ~ log(HomeRange_km2), data=MP2)
car::Anova(richmoddy2, type=3)

summary(richmoddy2)
plot(richmoddy2)

tab_model(richmoddy2)

hr = ggplot(MP2[-which(is.na(MP2$MSW93_Species)),], aes(x=log(HomeRange_km2), y=richness))+
  geom_point(aes(col=MSW93_Family, shape=MSW93_Order),size=4)+
  #geom_smooth(aes(col=MSW93_Family),method="lm",se=F)+
  geom_smooth(method="lm",color="black",size=1.5)+
  labs(x=bquote("log(Homerange (km"^2*"))"), y="Mean Individual Richness",col="Family",shape="Order")+
  annotate("text",x=-2.5, y=5.5,label=paste("R^2==0.22"),parse=T, size=5)+
  annotate("text",x=-2.5, y=5, label=paste("p==0.05"),parse=T, size=5)+
  theme_bw(base_size=14)

gp = ggplot(MP2[-which(is.na(MP2$MSW93_Species)),], aes(x=log(SocialGrpSize), y=richness))+
  geom_point(aes(col=MSW93_Family, shape=MSW93_Order),size=4)+
  #geom_smooth(aes(col=MSW93_Family),method="lm",se=F)+
  geom_smooth(method="lm",color="black",size=1.5,se=F,linetype="dotted")+
  labs(x=bquote("log(Social Group Size)"), y="Mean Individual Richness",col="Family",shape="Order")+guides(shape=F,col=F)+
  #annotate("text",x=-2.5, y=5.5,label=paste("R^2==0.27"),parse=T)+
  #annotate("text",x=-2.5, y=5, label=paste("p==0.03"),parse=T)+
  theme_bw(base_size=14)

bm = ggplot(MP2[-which(is.na(MP2$MSW93_Species)),], aes(x=log(Body_Mass_kg), y=richness))+
  geom_point(aes(col=MSW93_Family, shape=MSW93_Order),size=4)+
  #geom_smooth(aes(col=MSW93_Family),method="lm",se=F)+
  geom_smooth(method="lm",color="black",size=1.5,se=F,linetype="dotted")+
  labs(x=bquote("log(Body Mass (Kg))"), y="Mean Individual Richness",col="Family",shape="Order")+guides(shape=F,col=F)+
  #annotate("text",x=-2.5, y=5.5,label=paste("R^2==0.27"),parse=T)+
  #annotate("text",x=-2.5, y=5, label=paste("p==0.03"),parse=T)+
  theme_bw(base_size=14)

gridExtra::grid.arrange(gridExtra::grid.arrange(bm, gp,nrow=2), hr, ncol=2,widths=c(1,2))

```


### 2.5 Richness by species and location

```{r}

# subset data for Longitudinal species

# No zeros
LONGdata=RRA_present[RRA_present$Species %in% c("Warthog","Plains zebra","Elephant","Kudu","Buffalo","Impala"),]

# add rainfall
raindat = data.frame(Location=levels(as.factor(LONGdata$Location)), MAP2 = c(1336,894,576,769,541,529,1032,1006,900))

LONGdata=left_join(LONGdata,raindat)

richmodLONG = glm(richness ~ Species * Location, family="poisson", data=LONGdata)
richmodLONG = glm(richness ~ Species * MAP2, family="poisson", data=LONGdata)
#summary(richmodLONG)
car::Anova(richmodLONG)
richmodLONG2 = glm(richness ~ Species + Location, family="poisson", data=LONGdata)
richmodLONG2.1 = glm(richness ~ Species + MAP2, family="poisson", data=LONGdata)

richmodLONG3 = glm(richness ~ Species, family="poisson", data=LONGdata)
richmodLONG4 = glm(richness ~ Location, family="poisson", data=LONGdata)

bbmle::AICctab(richmodLONG2, richmodLONG2.1)
bbmle::AICtab(richmodLONG,richmodLONG2,richmodLONG3)
# no interaction is best

summary(richmodLONG2)
qqnorm(residuals(richmodLONG2))
testResiduals(simulateResiduals(richmodLONG2)) # marginal

# combined
MuMIn::r.squaredGLMM(richmodLONG2) # 62% combined

# marginal
MuMIn::r.squaredGLMM(richmodLONG3) # 60% only species
MuMIn::r.squaredGLMM(richmodLONG4) # 24% only location

car::Anova(richmodLONG2)

tab_model(richmodLONG2, transform=NULL, show.stat=T, show.se=T)

```


```{r}

### Now for data with zeros

LONGdata0=RRA2[RRA2$Species %in% c("Warthog","Plains zebra","Elephant","Kudu","Buffalo","Impala"),]

latdata=data.frame(Location=levels(as.factor(LONGdata0$Location)), lattitude=c(-18.787,-28.165, -18.964,-14.764,-23.861, 0.277, -12.861,-10.703,-2.577))

LONGdata0=left_join(LONGdata0,raindat)
LONGdata0=left_join(LONGdata0,latdata)

# Poisson models are overdispersed
# richmodLONG0 = glm(richness ~ Species * Location, family="poisson", data=LONGdata0)
# summary(richmodLONG0)
# car::Anova(richmodLONG0)
# richmodLONG20 = glm(richness ~ Species + Location, family="poisson", data=LONGdata0)
# richmodLONG30 = glm(richness ~ Species, family="poisson", data=LONGdata0)
# richmodLONG40 = glm(richness ~ Location, family="poisson", data=LONGdata0)
# 
# 
# bbmle::AICtab(richmodLONG0,richmodLONG20,richmodLONG30)
# # full interaction is best
# 
# qqnorm(residuals(richmodLONG0))
# testResiduals(simulateResiduals(richmodLONG0))
# summary(richmodLONG0) # over dispersion

# Even regular nb models are bad
# library(MASS)
# richmodLONG0 = glm.nb(richness ~ Species * Location, data=LONGdata0)
# summary(richmodLONG0)
# car::Anova(richmodLONG0)
# richmodLONG20 = glm.nb(richness ~ Species + Location, data=LONGdata0)
# richmodLONG30 = glm.nb(richness ~ Species, data=LONGdata0)
# richmodLONG40 = glm.nb(richness ~ Location,data=LONGdata0)
# bbmle::AICtab(richmodLONG0,richmodLONG20,richmodLONG30)
# # no interaction best
# 
# qqnorm(residuals(richmodLONG20))
# testResiduals(simulateResiduals(richmodLONG20))
# summary(richmodLONG20) # nope

LONGdata0$MAP = scale(LONGdata0$MAP2)
# check zero inflation

#richmodLONG0 = glmmTMB(richness ~ Species * Location, ziformula=~.,data=LONGdata0, family="nbinom2",REML=T)

# Full model
richmodLONG0 = glmmTMB(richness ~ Species + MAP + lattitude + (1|Location), ziformula=~., data=LONGdata0, family="nbinom2",REML=T)
# convergence problems

# Simplify
richmodLONG0 = glmmTMB(richness ~ Species + MAP + (1|Location), ziformula=~., data=LONGdata0, family="nbinom2",REML=T)
richmodLONG1 = glmmTMB(richness ~ Species + lattitude + (1|Location), ziformula=~., data=LONGdata0, family="nbinom2",REML=T)

bbmle::AICctab(richmodLONG0, richmodLONG1)
d = dredge(richmodLONG0) # this takes forever

d
# Species only is the best model (same as 98% data)

richmodLONGbest = glmmTMB(richness ~ Species + (1|Location), ziformula=~., data=LONGdata0, family="nbinom2",REML=T)

richmodLONGbest2 = glmmTMB(richness ~ Species + Location, ziformula=~., data=LONGdata0, family="nbinom2", REML=T)

sjstats::r2(richmodLONGbest)



lat = ggplot(LONGdata0, aes(x=lattitude, y=richness))+
  geom_jitter(aes(col=Species),alpha=0.5)+
  geom_smooth(aes(col=Species),method="lm",se=F)+
  geom_smooth(method="lm")+
  theme_bw(base_size=14)+guides(col=F)

mapp = ggplot(LONGdata0, aes(x=MAP2, y=richness))+
  geom_jitter(aes(col=Species),alpha=0.5)+
  geom_smooth(aes(col=Species),method="lm",se=F)+
  geom_smooth(method="lm")+
  theme_bw(base_size=14)+guides(col=F)

gridExtra::grid.arrange(lat,mapp, ncol=2)


longsp = LONGdata0 %>% group_by(Species, Location,MAP,MAP2,lattitude) %>% summarize_at(vars(richness),funs(mean))

ggplot(longsp, aes(x=lattitude, y=richness))+
  geom_point(aes(col=Species))+
  geom_smooth(method="lm")+
  geom_smooth(method="lm",aes(col=Species),se=F)

ggplot(longsp, aes(x=MAP2, y=richness))+
  geom_point(aes(col=Species))+
  geom_smooth(method="lm")+
  geom_smooth(method="lm",aes(col=Species),se=F)


```

### 2.6 Post hoc tests

```{r}
# no zeros
# emsLONG = emmeans(richmodLONG2,~ Species+Location, data=LONGdata)
# contrLONG = as.data.frame(contrast(emsLONG, adjust="holm"))
# more_info = as.data.frame(matrix(unlist(strsplit(as.character(contrLONG$contrast), split="[,]")), ncol=2, byrow=T))
# contrLONG$Species=more_info[,1]
# contrLONG$Location=more_info[,2]
# 
# contrLONG$Location=as.factor(contrLONG$Location)
# levels(contrLONG$Location)
# 
# newLoc = unlist(strsplit(as.character(contrLONG$Location),split=" "))
# contrLONG$Location = newLoc[-which(newLoc=="effect")]
# head(contrLONG)
# 
# contrLONG$Location = factor(contrLONG$Location,levels=c("Mpala","Kruger","Hwange","Kafue","Hluhluwe","Serengeti","Nyika","Niassa","Gorongosa"))
# 
# contrLONG$psymb = ifelse(contrLONG$p.value<0.05,"*","")
# 
# contrasts_present = ggplot(contrLONG, aes(x=reorder(Species,-estimate), y=estimate,group=Location))+
#   geom_point(aes(col=Location), position=position_dodge(width=0.6), size=2)+
#   geom_errorbar(aes(col=Location, ymin=estimate-(2*SE), ymax=estimate+(2*SE)), position=position_dodge(width=0.6), width=0.5, size=1.2)+
#   theme_bw(base_size=14)+
#   scale_color_sjplot(palette="ipsum")+
#   labs(x="", y="Effect Estimate")+
#   geom_hline(yintercept=0)+
#   geom_text(aes(label=psymb, y=estimate+(2*SE+0.05)), position=position_dodge(width=0.6))
# 


# now including 0s
emsLONGbest = emmeans(richmodLONGbest2,~ Species+Location, data=LONGdata0)
contrLONG0 = as.data.frame(contrast(emsLONGbest, adjust="holm"))
more_info0 = as.data.frame(matrix(unlist(strsplit(as.character(contrLONG0$contrast), split="[,]")), ncol=2, byrow=T))
contrLONG0$Species=more_info0[,1]
contrLONG0$Location=more_info0[,2]

newLoc0 = unlist(strsplit(as.character(contrLONG0$Location),split=" "))
contrLONG0$Location = newLoc0[-which(newLoc0=="effect")]

contrLONG0$Location=as.factor(contrLONG0$Location)

contrLONG0$Location = factor(contrLONG0$Location,levels=c("Mpala","Kruger","Hwange","Kafue","Hluhluwe","Serengeti","Nyika","Niassa","Gorongosa"))
contrLONG0=left_join(contrLONG0, raindat)
contrLONG0$Location = factor(contrLONG0$Location,levels=c("Mpala","Kruger","Hwange","Kafue","Hluhluwe","Serengeti","Nyika","Niassa","Gorongosa"))


contrLONG0$psymb = ifelse(contrLONG0$p.value<0.05,"*","")


contrast0 = ggplot(contrLONG0, aes(x=reorder(Species,-estimate), y=estimate,group=Location))+
  geom_point(aes(col=MAP2), position=position_dodge(width=0.6), size=2)+
  geom_errorbar(aes(col=MAP2, ymin=estimate-(2*SE), ymax=estimate+(2*SE)), position=position_dodge(width=0.6), width=0.5, size=1.2)+
  theme_bw(base_size=14)+
  scale_color_gradient(low="orange",high="blue")+
  labs(x="", y="Effect Estimate")+
  geom_hline(yintercept=0)+
  geom_text(aes(label=psymb, y=estimate+(2*SE+0.05)), position=position_dodge(width=0.6))

#gridExtra::grid.arrange(contrasts_present,contrast0)

contrast0

```

```{r}

# plot with actual data points?
longsum=Rmisc::summarySE(LONGdata, groupvars=c("Location","Species"), measurevar="richness")

LONGdata %>% group_by(Species, Location) %>% summarize(n())
ggplot(LONGdata,  aes(x=MAP2, y=richness))+
  geom_jitter(aes(col=Species))+
  geom_smooth(aes(col=Species), method="lm", se=F)+
  geom_smooth(method="lm")
```


## Species-level richness

```{r}
specpool(MPdata_0[,15:501],MPdata_0$Species)


MPdata_0 %>% group_by(Species) %>% summarize(n())
# need to randomly select rows for species-level richness presumably?

newMP = subset(MPdata_0, Species==levels(as.factor(MPdata_0$Species))[1])
for(i in 2:length(levels(as.factor(MPdata_0$Species)))){
  sp = subset(MPdata_0, Species==levels(as.factor(MPdata_0$Species))[i])
  n = dim(sp)[1]
  if(n > 15){
    subsp = sp[sample(1:n, 15),]
  }else{
    subsp=sp[1:n,]
  }
  newMP = rbind(newMP, subsp)
}

newMP %>% group_by(Species) %>% summarize(n())

# camel has 6, hippo 9, kudu 3, oryx 7, waterbuck 2

newMP_sp = newMP %>% group_by(Species) %>% summarize_at(vars(MOTU_1:MOTU_1677), funs(sum))
newMP_sp$richness = specnumber(newMP_sp[,-1])

ggplot(newMP_sp, aes(x=reorder(Species,-richness), y=richness))+
  geom_col(aes(fill=Species))+
  guides(fill=F)+
  theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0))+
  labs(y="Richness across ~15 individuals")

speculative = specpool(newMP[,15:501],newMP$Species)
speculative$Host=row.names(speculative)


colorsp = data.frame(Host=levels(MPdata$Species), colors=animal_colors)
colorsp$colors[which(colorsp$Species %in% speculative$Host)]
speculative=left_join(speculative, colorsp)

ggplot(speculative, aes(x=reorder(Host,-boot), y=boot))+
  geom_point( size=3, color=as.character(speculative$colors))+
  geom_errorbar(aes(ymin=boot-2*boot.se, ymax=boot+2*boot.se), width=0.25, size=1.5,color=as.character(speculative$colors))+
  geom_point(aes(x=reorder(Host, -boot), y=jack1), position=position_nudge(x=0.25), size=3,color=as.character(speculative$colors))+
  geom_errorbar(aes(x=reorder(Host, -boot), ymin=jack1-2*jack1.se, ymax=jack1+2*jack1.se),color=as.character(speculative$colors), position=position_nudge(x=0.25), width=0.25, size=1.5)+
  guides(col=F)+
  #scale_color_manual(values=as.character(colorsp$colors[which(colorsp$Species %in% speculative$Host)]))+
  theme_bw(base_size=14)+
  theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0))+
  labs(y="Estimated Total Parasite OTUs", x="Host Species")


ggplot(speculative, aes(x=reorder(Host,-jack1), y=jack1))+
  geom_point(size=3, color=as.character(speculative$colors))+
  geom_errorbar(aes(ymin=jack1-2*jack1.se, ymax=jack1+2*jack1.se), width=0.25, size=1.5,color=as.character(speculative$colors))+
  geom_point(aes(x=reorder(Host, -jack1), y=Species), size=3,color=as.character(speculative$colors), shape=15)+
  guides(col=F)+
  theme_bw(base_size=14)+
  theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0))+
  geom_text(aes(x=Host, y=Species-5, label=n))+
  labs(y="Estimated Total Parasite Species", x="Host Species")



```

### Specpool for location data

```{r}

# need species and location 
LONGdata0$sp_loc=paste(LONGdata0$Species,LONGdata0$Location, sep="_")

speculative_long = specpool(LONGdata0[,15:189],LONGdata0$sp_loc)
speculative_long$sp_loc=row.names(speculative_long)

new=as.data.frame(matrix(unlist(strsplit(speculative_long$sp_loc, split="_")),ncol=2,byrow=T))
names(new)=c("Host","Location")
speculative_long$Host = new$Host
speculative_long$Location = new$Location

ggplot(speculative_long, aes(x=reorder(Host,-boot), y=boot))+
  geom_point( size=3, aes(color=Location), position=position_dodge(width=0.5))+
  geom_errorbar(aes(ymin=boot-2*boot.se, ymax=boot+2*boot.se, color=Location), width=0.25, size=1.5, position=position_dodge(width=0.5))+
  geom_point(aes(x=reorder(Host, -boot), y=jack1, col=Location), position=position_dodge(width=0.5), size=3, shape=2)+
  geom_errorbar(aes(x=reorder(Host, -boot), ymin=jack1-2*jack1.se, ymax=jack1+2*jack1.se, col=Location), position=position_dodge(width=0.5), width=0.25, size=1, linetype="dotted")+
  guides(col=F)+
  theme_bw(base_size=14)+
  theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0))+
  labs(y="Estimated Total Parasite OTUs", x="Host Species")

```


### Specpool for literature

```{r}

litotu = read.csv("host_parasite_lit.csv")
head(litotu)

# need to group by the source also
litlong = litotu %>% gather(R_1:R_12, key=refID, value=Ref)
# remove blanks
litlong=litlong[-which(is.na(litlong$Ref)),]
litlong=litlong[-which(litlong$Ref==""),]
# this gives the number of parasite species per reference for each host
litlong2 = litlong %>% group_by(Ref, Host) %>% summarise(n())
litlong3 = litlong %>% group_by(Ref, Host, Species_clean) %>% summarise(N=n())
dim(unique(litlong3))
litlong4 = litlong3 %>% spread(key=Species_clean, value=N, fill=0)

totspecs = specpool(litlong4[-c(1,2)], pool=litlong4$Host) # I did not keep track of every single reference for cows and donkeys

# from below, the richness for cows is 87 and for donkeys it is 80
totspecs[c(3),] = c(87,87,0,87,0,87,87,0,NA)
totspecs[c(5),] = c(80,80,0,80,0,80,80,0,NA)

totspecs$Host=c("Impala","Hartebeest","Cattle","Camel","Donkey","Plains zebra","Grevys zebra","Giraffe","Elephant","DikDik","Grants gazelle","Oryx","Warthog","Buffalo","Eland")


totspecs=left_join(totspecs, colorsp)
totspecs$colors[3]='#6A5ACD'

ggplot(totspecs, aes(x=reorder(Host,-jack2), y=boot))+
  geom_point( size=3, color=as.character(totspecs$colors))+
  geom_errorbar(aes(ymin=boot-2*boot.se, ymax=boot+2*boot.se), width=0.25, size=1.5,color=as.character(totspecs$colors))+
  geom_point(aes(x=reorder(Host, -boot), y=jack2), position=position_nudge(x=0.25), size=3,color=as.character(totspecs$colors))+
  geom_errorbar(aes(x=reorder(Host, -boot), ymin=jack2-2*jack1.se, ymax=jack2+2*jack1.se),color=as.character(totspecs$colors), position=position_nudge(x=0.25), width=0.25, size=1.5)+
  guides(col=F)+
  #scale_color_manual(values=as.character(colorsp$colors[which(colorsp$Species %in% speculative$Host)]))+
  theme_bw(base_size=14)+
  theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0))+
  labs(y="Estimated Total Parasite Species", x="Host Species")


ggplot(totspecs, aes(x=reorder(Host,-jack2), y=jack2))+
  geom_point( size=3, color=as.character(totspecs$colors))+
  geom_errorbar(aes(ymin=jack2-2*jack1.se, ymax=jack2+2*jack1.se), width=0.25, size=1.5,color=as.character(totspecs$colors))+
  geom_point(aes(x=reorder(Host, -boot), y=Species), size=3,color=as.character(totspecs$colors), shape=15)+
  guides(col=F)+
  theme_bw(base_size=14)+
  theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0))+
  geom_text(aes(x=Host, y=Species-5, label=n))+
  labs(y="Estimated Total Parasite Species", x="Host Species")




ggplot(totspecs, aes(x=Host, y=boot))+
  geom_point()


# duplicate IDs
litotu2 = litotu %>% group_by(Host, Species_clean) %>% summarize_at(vars(Refs), funs(sum))
litotu2 = litotu2 %>% spread(key=Species_clean, value=Refs, fill=0)
litotu2$total_refs = rowSums(litotu2[,-1])
litotu2$richness = specnumber(litotu2[,-c(1,298)])
litotu2

ggplot(litotu2, aes(x=total_refs, y=richness))+
  geom_point(aes(color=Host),size=3)+
  geom_smooth(method="lm", formula=y~log(x))+
  scale_y_continuous(limits=c(0,100))+
  theme_bw(base_size=14)+
  labs(x="Total Records", y="Parasite Richness")


compys = left_join(speculative, totspecs, by="Host")
head(compys)
ggplot(compys, aes(x=Species.x, y=Species.y))+
  geom_point(aes(color=Host), size=3)+
  geom_hline(yintercept=mean(compys$Species.y,na.rm=T))+
  geom_vline(xintercept=mean(compys$Species.x))+
  labs(x="OTU richness",y="Literature richness")

plottot = ggplot(compys, aes(x=Species.x, y=Species.y))+
  geom_point(aes(color=Host), size=3)+
  geom_abline(intercept=0, slope=1)+
  labs(x="OTU richness",y="Literature richness")+
  theme_bw(base_size=14)#+
  guides(col=F)


ggplot(compys, aes(x=jack1.x, y=jack1.y))+
  geom_point(aes(color=Host), size=3)+
  geom_hline(yintercept=mean(compys$Species.y,na.rm=T))+
  geom_vline(xintercept=mean(compys$Species.x))+
  labs(x="OTU richness",y="Literature richness")

plotspectot = ggplot(compys, aes(x=jack1.x, y=jack1.y))+
  geom_point(aes(color=Host), size=3)+
  geom_abline(intercept=0, slope=1)+
  labs(x="OTU richness",y="Literature richness")+
  theme_bw(base_size=14)
plotspectot

gridExtra::grid.arrange(plottot, plotspectot, ncol=2)
```



## 3 Connecting to Vanessa's data

Not sure if this is necessary or helpful really.

```{r}

# # she also included zeros for richness
# vanessa = read.csv("Vanessa_data.csv")
# head(vanessa)
# str(vanessa)
# names(vanessa)[1]=c("Species")
# vanessa$Species=as.character(vanessa$Species)
# vanessa$Species = str_trim(vanessa$Species)
# numR=numR[-is.na(numR$Location),]
# vanessa$Species = as.factor(vanessa$Species)
# levels(as.factor(numR$Species))
# levels(vanessa$Species)[c(3,6)]=c("DikDik","Grants gazelle")
# 
# vancomp =left_join(numR,vanessa)
# 
# head(vancomp)
# vancomp = vancomp[-which(is.na(vancomp$Total.richness)),]
# dim(vancomp)
# names(vancomp)=c("Location","Species","N_OTU","totRich_OTU","indRich_OTU","N_Van","N2_Van","S_prev","C_prev","indRich_Van","totRich_Van")
# 
# vanmcomp=vancomp[vancomp$Location=="Mpala",]
# 
# # recalculate Vanessas numbers excluding 0s
# str(vanmcomp)
# vanmcomp$N_Van = as.numeric(as.character(vanmcomp$N_Van))
# vanmcomp$N2_Van = as.numeric(as.character(vanmcomp$N2_Van))
# 
# vanmcomp$nonzero = ((vanmcomp$S_prev/100 * vanmcomp$N2_Van))
# vanmcomp$new_Rich_van = vanmcomp$indRich_Van * vanmcomp$N2_Van / vanmcomp$nonzero
# 
# # recalculate without coccidia
# vanmcomp$cnumber = vanmcomp$C_prev/100 * vanmcomp$N2_Van
# vanmcomp$possrich = (vanmcomp$new_Rich_van * vanmcomp$nonzero - vanmcomp$cnumber)/ vanmcomp$nonzero
# 
# ggplot(vanmcomp, aes(x=possrich, y=indRich_OTU))+
#   geom_point(aes(color=Species, size=C_prev))+
#   geom_smooth(method="lm",se=F)
# 
# 
# 
# # very low power without using the other species....
# cor.test(vanmcomp$indRich_OTU, vanmcomp$possrich, method="spearman")
# 
# vanmcomp
# 
# # This would require a randome subsample to control for size
# ggplot(vanmcomp, aes(x=totRich_Van, y=totRich_OTU))+
#   geom_point(aes(color=Species, size=N_OTU))+
#   geom_smooth(method="lm",se=F)
# 
# 
# cor.test(vanmcomp$totRich_Van, vanmcomp$totRich_OTU)

```


### 4 Comparing host range among MOTUs

```{r}

MOTU = RRA2 %>% group_by(Species) %>%
  summarize_at(vars(MOTU_1:MOTU_1677),funs(sum(.))) %>%
  as.data.frame()%>%
  .[,-1] %>%
  t()%>%
  as.data.frame()
names(MOTU) = levels(as.factor(RRA2$Species))
MOTU = MOTU %>% mutate_at(vars(`Grants gazelle`:Kudu), funs(ifelse(.>0,1,0)))
MOTU$Hosts = rowSums(MOTU)
MOTU$MOTU = names(RRA2)[15:501]

ggplot(MOTU, aes(x=Hosts, y=reorder(MOTU,Hosts)))+
geom_jitter(width=0.1,height=0.1,alpha=0.5)+
scale_x_continuous(breaks=c(1,2,3,4,5,6,7,8,9,10,11,12,13,14))

hstct = data.frame(host = names(colSums(MOTU[,-c(20,21)])), count=colSums(MOTU[,-c(20,21)]))
ggplot(hstct, aes(x=count, y=reorder(host,count)))+
  geom_jitter()
```


```{r}
# bring in MOTU information
motuinfo = read.delim("nematodes_2018_assembled_setid_filterE2_nl_uniq_c10_ann_assignEMBL_nematoda_sumatra_sumaclust_assignedR134.tab")
head(motuinfo)
motu98 = data.frame(cluster = (motuinfo$cluster_sumatra99), name=motuinfo$scientific_name_ok)

motu98 = unique(motu98)

motu98$MOTU = paste("MOTU_",motu98$cluster, sep="")
motu98
MOTU = left_join(MOTU, motu98)
MOTU$name2 = MOTU$name

# Super family -- might be worth doing family instead
levels(MOTU$name) = c("Ancyclostomatidae","Ancyclostomatidae","Ancyclostomatidae","C.elegans","Cloacinidae","Cooperiidae", "Strongylidae","Strongylidae","Haemonchidae","Trichostrongylidae","Ancyclostomatidae","Chabertiidae","Chabertiidae","Protostrongylidae","Strongylida (order)","Strongylidae", "Strongyloidea (super family)","Strongylidae","Haemonchidae","Trichostrongylidae","Trichostrongylidae","Trichostrongyloidea (super family)","Trichostrongylidae","Trichostrongylidae", "Trichostrongylidae", "Protostrongylidae")


  ggplot(MOTU, aes(x=Hosts, y=reorder(MOTU,Hosts)))+
  geom_jitter(width=0.1,height=0.1,alpha=0.5, aes(color=name), size=2)+
    scale_x_continuous(breaks=c(1,2,3,4,5,6,7,8,9,10))+
    scale_y_discrete(labels=NULL)+
    theme_bw(base_size=14)+facet_wrap(~name)+guides(col=F)
  
```



## 5. Ordination

### 5.1 Create ordination objects using Jaccard distance

```{r}

# remove rows and columns with NA values, although these rows are technically 0s
names(MPdata)

# Elephant issue
#ordMRC = metaMDS(MPdata[,15:189], k=3, binary=T,noshare=T,trymax=100)
MPdata_noEle = MPdata[MPdata$Species!="Elephant",]
# disconnected points
#MPdata_noEle[which(ordMRC$points[,2]<0),]
MPdata_ok = MPdata_noEle[-c(17,22,27),]
ordMRC = metaMDS(MPdata_ok[,15:501], k=3, binary=T, distance="jaccard", noshare=T, trymax=100, trace=0)
ordMRC$converged

# subset data for Longitudinal species
LONGdata=RRA_present[RRA_present$Species %in% c("Warthog","Plains zebra","Elephant","Kudu","Buffalo","Impala"),c(1:5,15:501)]

# remove any columns that now sum to 0
LONGdata= LONGdata[,-(which(colSums(LONGdata[,-c(1:5)])==0)+5)]
dim(LONGdata)

ordLONG=metaMDS(LONGdata[,6:389], distance="jaccard", binary=T, k=3, noshare=T, trymax=100, trace=0)

#may need to run for more iterations
ordLONG$converged
ordLONG=metaMDS(LONGdata[,6:389], distance="jaccard", binary=T, k=3, noshare=T, trymax=100,previous.best=ordLONG,trace=0)

# takes some time
# two points were disconnected and have been excluded here:
ord = metaMDS(RRA_present[-c(43,71),15:501], distance="jaccard",binary=T, k=3, noshare=T, trymax=100, trace=0)
ord$converged
ord = metaMDS(RRA_present[,15:501], distance="jaccard",binary=T, k=3, noshare=T, trymax=100, trace=0, previous.best = ord)
```

Plot ordinations

```{r}

plot(ordMRC)

plot(ordLONG)

plot(ord)

```


### 5.2 Examine species and location variables

```{r}
# Construct data frames with this information
names(MPdata_ok)
envMRC=MPdata_ok[,c(3,4)]
# check NAs
envMRC$Period[which(is.na(envMRC$Period))]

# Conduct a factor fit of species and sampling period
factorfit(ordMRC$points,envMRC,1000) #crashed
# Using adonis methods
MRCdist = vegdist(MPdata_ok[,15:501], distance="jaccard")
MRCdiv = adonis2(MRCdist ~ Species+Period,  data=envMRC, permutations=999, method="jaccard")
MRCdiv
# Different R2 values and period is less important than in factorfit.



# For longitudinal data
envLONG=LONGdata[,c(2,3)]

# Conduct a factor fit of species and location
factorfit(ordLONG$points, envLONG,1000)
# Using adonis methods
LONGdist = vegdist(LONGdata[,6:389], distance="jaccard")
LONGdiv=adonis2(LONGdist ~ Species*Location, data=envLONG, permutations=999, method="jaccard")
LONGdiv
LONGdiv$Df


# all data
ALLdist = vegdist(RRA_present[-c(43,71),15:501], distance="jaccard")
envALL = RRA_present[-c(43,71),c(2,3)]
ALLdiv=adonis2(ALLdist ~ Species*Location, data=envALL, permutations=999, method="jaccard")
ALLdiv


#write.csv(RRA_present, "RRA_present_99.csv",row.names=F)

```

### 5.3 More plotting

Better plots with parasite info are displayed at the bottom

```{r}
# Plotting MRC
plotMDSMRC = data.frame(species=envMRC[,1], period=envMRC[,2],x1=ordMRC$points[,1], x2=ordMRC$points[,2], x3=ordMRC$points[,3])

# Plotting longitudinal data
plotMDSLONG = data.frame(location=envLONG[,1], species=envLONG[,2], x1=ordLONG$points[,1], x2=ordLONG$points[,2], x3=ordLONG$points[,3])


ggplot(plotMDSMRC, aes(x=x1, y=x2))+
  geom_jitter(aes(col=Period), size=2, width=0.1, height=0.1, alpha=0.5, stroke=1.5)+
  geom_convexhull(aes(fill=Period), alpha=0.3)+
  facet_wrap(~Species)+
  theme_gray(base_size=12)+
  scale_color_sjplot(palette="warm")+
  scale_fill_sjplot(palette="warm")

ggplot(plotMDSMRC, aes(x=x1, y=x2))+
  geom_jitter(aes(col=Species), size=2, width=0.1, height=0.1, alpha=0.5, stroke=1.5)+
  geom_convexhull(aes(fill=Species), alpha=0.3)+
  theme_gray(base_size=12)+
  scale_color_sjplot(palette="warm")+
  scale_fill_sjplot(palette="warm")


ggplot(plotMDSMRC, aes(x=x2, y=x3))+
  geom_jitter(aes(col=Species), size=2, width=0.1, height=0.1, alpha=0.5, stroke=1.5)+
  geom_convexhull(aes(fill=Species), alpha=0.3)+
  theme_gray(base_size=12)+
  scale_color_sjplot(palette="ipsum")+
  scale_fill_sjplot(palette="ipsum")


ggplot(plotMDSLONG, aes(x=x1, y=x2))+
  geom_jitter(aes(col=Species), size=2, width=0.1,height=0.1, alpha=0.5,stroke=1.5)+
  #scale_shape_manual(values=1:nlevels(plotMDSLONG$location))+
  geom_convexhull(aes(fill=Species),alpha=0.5)+
  facet_wrap(~Location)+
  theme_gray(base_size=14)+
  scale_color_sjplot(palette="warm")+
  scale_fill_sjplot(palette="warm")


ggplot(plotMDSLONG, aes(x=x1, y=x2))+
  geom_jitter(aes(col=species, shape=location), size=2, width=0.1,height=0.1, alpha=0.5,stroke=1.5)+
  scale_shape_manual(values=1:nlevels(plotMDSLONG$location))+
  geom_convexhull(aes(fill=species),alpha=0.5)+
  theme_bw(base_size=14)+
  scale_color_manual(values=as.character(colorsp$colors[which(colorsp$Host %in% c("Plains zebra","Elephant","Buffalo","Warthog","Kudu","Impala"))]))+
  scale_fill_manual(values=as.character(colorsp$colors[which(colorsp$Host %in% c("Plains zebra","Elephant","Buffalo","Warthog","Kudu","Impala"))]))


plotMDSLONG$location = factor(plotMDSLONG$location, levels=c("Mpala","Kruger","Hwange","Kafue","Hluhluwe","Serengeti","Nyika","Niassa","Gorongosa"))

ggplot(plotMDSLONG, aes(x=x1, y=x2))+
  geom_jitter(aes(col=species, shape=location), size=2, width=0.1,height=0.1, alpha=0.5,stroke=1.5)+
  scale_shape_manual(values=1:nlevels(plotMDSLONG$location))+
  geom_convexhull(aes(fill=species,linetype=location),alpha=0.5)+
  theme_bw(base_size=18)+
  scale_color_manual(values=as.character(colorsp$colors[which(colorsp$Host %in% c("Plains zebra","Elephant","Buffalo","Warthog","Kudu","Impala"))]))+
  scale_fill_manual(values=as.character(colorsp$colors[which(colorsp$Host %in% c("Plains zebra","Elephant","Buffalo","Warthog","Kudu","Impala"))]))+
  #facet_wrap(~location)+
  labs(x="MDS1",y="MDS2", color="Host Species",fill="Host Species")+
  guides(shape=F,linetype=F)



```

MRC notes: Separating species and season seems to show a fair amount of consistency over time. There don't appear to be marked seasonal differences, at least at this scale. However, I can definitely see that there is something weird going on with one of the hippo samples from 2017 June. It looks to be very squarely in the equid realm...


### 5.4 Weird observations

```{r}
#RRA_present[RRA_present$Species=="Hippo",]

# MRC_17_HIP_93 is the problem sample -- is there any chance at all that it is actually zebra?

# These are weird from longitudinal set
LONGdata[which(plotMDSLONG$Species=="Buffalo"&plotMDSLONG$Location=="Serengeti"&plotMDSLONG$x1>0),]
```


One thing that seems a little weird to me is that the serengeti buffalo appear to have some MOTUs that overlap with plains zebra. Might need to check those out more in depth.

### Geographical distance and similarity

```{r}

locdists = read.csv("location_distances.csv")

upprLONG = upper.tri(LONGdist, diag=F)
LONGdistdf = as.matrix(LONGdist)

### Make function to calculate this
infraspecific_jaccard_comp = function(x){
    warts = LONGdistdf[which(LONGdata$Species==x),which(LONGdata$Species==x)]
    wart2 = as.dist(warts)
    uprWART = upper.tri(wart2, diag=F)
    wartcomps = warts[uprWART]
    locationvec = LONGdata$Location[which(LONGdata$Species==x)]
    locations_long = data.frame(location1 = "Hluhluwe",location2="Hluhluwe")
    for(i in 1:dim(uprWART)[2]){
      for(j in 1:dim(uprWART)[2]){
        location1 =locationvec[i]
        location2 =locationvec[i+j]
        locationadd = data.frame(location1=location1, location2=location2)
        locations_long=rbind(locations_long,locationadd)
      }
    }
    locations_long = (locations_long[-which(is.na(locations_long$location2)),])
    locations_long = locations_long[-1,]
    locations_long$jaccard = wartcomps
    locations_long$both_locs = paste(locations_long$location1,locations_long$location2, sep="_")
    locations_long2 = left_join(locations_long,locdists)
    return(locations_long2)
}

# Takes a minute
warthog_long = infraspecific_jaccard_comp(x="Warthog")
buffalo_long = infraspecific_jaccard_comp(x="Buffalo")
impala_long = infraspecific_jaccard_comp(x="Impala")
plains_long = infraspecific_jaccard_comp(x="Plains zebra")
elephant_long = infraspecific_jaccard_comp(x="Elephant")
kudu_long = infraspecific_jaccard_comp(x="Kudu")
  

```


```{r}

warthog_long$species = rep("Warthog",length(warthog_long$location1))
buffalo_long$species = rep("Buffalo",length(buffalo_long$location1))
impala_long$species = rep("Impala",length(impala_long$location1))
plains_long$species = rep("Plains zebra",length(plains_long$location1))
elephant_long$species = rep("Elephant",length(elephant_long$location1))
kudu_long$species = rep("Kudu",length(kudu_long$location1))

alllocs_long = rbind(warthog_long,buffalo_long,impala_long,plains_long,elephant_long,kudu_long)

ggplot(alllocs_long, aes(x=Distance, y=log(1-jaccard)))+
  geom_jitter(aes(col=species), alpha=0.1,width=10)+
  geom_smooth(method="lm", aes(col=species), se=F, size=1.2)+
  geom_smooth(method='lm',size=1.5, color="black")+
  scale_color_manual(values = as.character(newcols$colors[c(16,4,11,15,3,18)]))+
  theme_bw(base_size=14)+
  labs(x="Distance (km)",y="Jaccard Index",color="Host Species")+
  ggtitle("Infra-community")


# could bootstrap
alllocs_long$species = factor(alllocs_long$species, levels=rev(levels(as.factor(alllocs_long$species))))
alllocs_long$distsmall=alllocs_long$Distance/1000
anothermod = lmer(jaccard~Distance*species+(1|both_locs),data=alllocs_long)
summary(anothermod)
car::Anova(anothermod)
testResiduals(simulateResiduals(anothermod))

```




```{r}
### now for overall community
# group LONG data by species and location and then calculate distance matrix
longsmall = LONGdata %>% group_by(Location, Species) %>% summarize_at(vars(MOTU_1:MOTU_1677), funs(mean))

# don't include loactions with <8
longn = LONGdata %>% group_by(Location, Species) %>% summarize(N = n())
longsmall = left_join(longsmall, longn)
longsmall = longsmall[-which(longsmall$N<8),] # about half as many

longdist_small = vegdist(longsmall[,-c(1,2,387)], distance="jaccard")
upprLONGsmall = upper.tri(longdist_small, diag=F)
LONGdistdfsmall = as.matrix(longdist_small)

infraspecific_jaccard_comp2 = function(x){
    warts = LONGdistdfsmall[which(longsmall$Species==x),which(longsmall$Species==x)]
    wart2 = as.dist(warts)
    uprWART = upper.tri(wart2, diag=F)
    wartcomps = warts[uprWART]
    locationvec = longsmall$Location[which(longsmall$Species==x)]
    locations_long = data.frame(location1 = "Hluhluwe",location2="Hluhluwe")
    for(i in 1:dim(uprWART)[2]){
      for(j in 1:dim(uprWART)[2]){
        location1 =locationvec[i]
        location2 =locationvec[i+j]
        locationadd = data.frame(location1=location1, location2=location2)
        locations_long=rbind(locations_long,locationadd)
      }
    }
    locations_long = (locations_long[-which(is.na(locations_long$location2)),])
    locations_long = locations_long[-1,]
    locations_long$jaccard = wartcomps
    locations_long$both_locs = paste(locations_long$location1,locations_long$location2, sep="_")
    locations_long2 = left_join(locations_long,locdists)
    return(locations_long2)
}


kudus = infraspecific_jaccard_comp2(x="Kudu")
buffs = infraspecific_jaccard_comp2(x="Buffalo")
elephants = infraspecific_jaccard_comp2(x="Elephant")
impala=infraspecific_jaccard_comp2(x="Impala")
plains = infraspecific_jaccard_comp2(x="Plains zebra")
warthogs = infraspecific_jaccard_comp2(x="Warthog")

kudus$species = rep("Kudu",length(kudus$location1))
buffs$species = rep("Buffalo",length(buffs$location1))
elephants$species = rep("Elephant",length(elephants$location1))
impala$species = rep("Impala",length(impala$location1))
plains$species =rep("Plains zebra",length(plains$location1))
warthogs$species =rep("Warthog",length(warthogs$location1))

alllocs_small = rbind(kudus, buffs, elephants, impala, plains, warthogs)

ggplot(alllocs_small, aes(x=Distance, y=log(1-jaccard)))+
  geom_point(aes(col=species),alpha=0.8,size=2)+
  geom_smooth(method="lm",se=F, aes(col=species), size=1.2,alpha=0.5)+
  theme_bw(base_size=14)+
    geom_smooth(method="lm", size=1.5, color="black")+
    scale_color_manual(values = as.character(newcols$colors[c(16,4,11,3,18)]))+
  labs(x="Distance (km)", y="Jaccard Index", col="Host Species")+
  ggtitle("Component Community")


modeltest2 = lm(jaccard ~ species*Distance, data=alllocs_small)
step(modeltest2)

```


### 5.5 Measuring Dispersion

```{r}

dispersion<-betadisper(MRCdist, group=envMRC$Species)
permutest(dispersion)

plot(dispersion, hull=FALSE, ellipse=TRUE, label.cex=0.5)


dispersion=betadisper(LONGdist, group=envLONG$Species)
permutest(dispersion)

plot(dispersion, hull=FALSE, ellipse=TRUE, label.cex=0.5)


```


## 6. Detecting Phylogenetic signal

```{r}

tree = read.tree("mammaltree.txt")

# Use ordination results to find phylogenetic signal
# using MRC data
ALLPOINTS = ord$points
MDSdf = data.frame(species=RRA_present$Species[-c(43,71)], location=RRA_present$Location[-c(43,71)], x1 = ALLPOINTS[,1], x2 =ALLPOINTS[,2], x3=ALLPOINTS[,3])
ALLRRA = MDSdf %>% group_by(species) %>% summarize_at(vars(x1:x3), funs(mean))

#MRCFF = factorfit(ordMRC$points,envMRC,1000)
#avgRRA = MRCFF$centroids[1:18,]
#RRA_sp = RRA_present %>% group_by(Species) %>% summarize_at(vars(richness),funs(mean))

RRA_sp0 = RRA2 %>% group_by(Species) %>% summarize_at(vars(richness), funs(mean))

mydat=left_join(RRA_sp0,ALLRRA, by=c("Species"="species"))

row.names(mydat)=c("Gazella_granti","Loxodonta_africana","Equus_grevyi","Oryx_gazella","Equus_burchellii","Alcelaphus_buselaphus","Equus_asinus","Aepyceros_melampus","Equus_hybrid","Camelus_dromedarius","Taurotragus_oryx","Phacochoerus_africanus","Madoqua_kirkii","Giraffa_camelopardalis","Kobus_ellipsiprymnus","Hippopotamus_amphibius","Syncerus_caffer","Bos_taurus","Tragelaphus_strepsiceros")

nt = extractTree(tree)
nt
snt=(subset(nt, tips.exclude="Tragelaphus_scriptus"))

# need to reorder my dat according to tree order
mydat = as.data.frame(mydat)
mydat$spname = row.names(mydat)
mydat = mydat[c(4,6,8,1,15,13,18,17,11,19,14,16,12,10,7,5,3,2),]

p4d = phylo4d(snt, mydat[,-c(1,6)])

barplot.phylo4d(p4d, tree.type="phylo", tree.ladderize = T, scale=F, center=F)

PSig = phyloSignal(p4d=p4d, method="all")
PSig

# This takes a second
rich=phyloCorrelogram(p4d,trait="richness")
plot(rich)
axis1 = phyloCorrelogram(p4d, trait="x1")
plot(axis1)
axis2 = phyloCorrelogram(p4d, trait="x2")
plot(axis2)
axis3 = phyloCorrelogram(p4d, trait="x3")
plot(axis3)

nem.lipa=lipaMoran(p4d)
nem.lipa.p4d = lipaMoran(p4d,as.p4d=T)

barplot.phylo4d(nem.lipa.p4d, bar.col=(nem.lipa$p.value <0.05)+1, center=T, scale=F)
barplot.phylo4d(p4d, bar.col=(nem.lipa$p.value <0.05)+1, center=T, scale=F)

```

Richness is explained by phylogeny only for cattle/buffalo and plains/grevys
MDS coordinates are explained by phylogeny for MDS 2 and 3. It seems that most of the variation on MDS 1 describes Elephants vs. Hippos. 
Could narrow this to only bovids for example...

Could also look at specific MOTUs or MOTUs grouped by family.


```{r}

# Extract specific MOTUS and determine weights on MDS
motu_wts = as.data.frame(ord$species)
motu_wts$MOTU =row.names(motu_wts)
names(motu_wts)
merge(motu_wts, MOTU[,c(20:22,24)])
motu_wts = unique(left_join(motu_wts, MOTU[,c(20:22,24)]))
motu_wts$name= motu_wts$name2
#View(motu_wts)
motu_wts$name = as.factor(motu_wts$name)

levels(motu_wts$name) # name and name 2 are off

levels(motu_wts$name) = c("Ancyclostomatidae","Ancyclostomatidae","Ancyclostomatidae","C.elegans","Cloacinidae","Cooperiidae", "Strongylidae","Strongylidae","Haemonchidae","Trichostrongylidae","Ancyclostomatidae","Chabertiidae","Chabertiidae","Protostrongylidae","Strongylida (order)","Strongylidae", "Strongyloidea (super family)","Strongylidae","Haemonchidae","Trichostrongylidae","Trichostrongylidae","Trichostrongyloidea (super family)","Trichostrongylidae","Trichostrongylidae", "Trichostrongylidae", "Protostrongylidae")

motu_wts %>% gather(MDS1:MDS3, key=Axis, value=Score)%>%
ggplot(aes(x=Axis, y=Score))+
  geom_point(aes(group=name,col=name), position=position_dodge(width=0.75))+
  theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0))


# task -- find centroid for each parasite family.
psitecent = motu_wts %>% group_by(name)%>% summarize_at(vars(MDS1:MDS3), funs(mean(.,na.rm=T)))
#plotMDSMRC2 = left_join(plotMDSMRC, colorsp, by=c("Species"="Host"))
plotMDSMRC2 = left_join(ALLRRA, colorsp, by=c("species"="Host"))

plotMDSMRC2$species = factor(plotMDSMRC2$species, levels=levels(reorder(RRA2$Species,-RRA2$richness)))
MDSdf$species = factor(MDSdf$species, levels =levels(plotMDSMRC2$species))

levels(as.factor(plotMDSMRC2$species))
levels(plotMDSMRC2$colors)= c("dodgerblue", "green3","magenta3","cadetblue1","maroon1",  "dodgerblue3","maroon3","deepskyblue3", "maroon", "gold3", "dodgerblue4","green4","deepskyblue", "gold", "lightsteelblue3", "greenyellow",  "darkorchid4","darkorchid1", "lightskyblue2")

ggplot(plotMDSMRC2, aes(x=species, y=x1))+
  geom_point(color=as.character(plotMDSMRC2$colors))

psitecent %>%
  ggplot(aes(x=MDS1, y=MDS2))+
  theme_gray(base_size=12)+
  geom_jitter(data=MDSdf, aes(x=x1,y=x2,col=species), size=2, stroke=1.5,width=0.1, height=0.1, alpha=0.5)+
  geom_convexhull(data=MDSdf, aes(x=x1,y=x2, fill=species), alpha=0.3)+
  scale_fill_manual(values=as.character(levels(plotMDSMRC2$colors)))+
  scale_color_manual(values=as.character(levels(plotMDSMRC2$colors)))+
  geom_segment(aes(xend=0, yend=0))+
  geom_text(aes(label=name))+
  theme_bw(base_size=14)



psitecent %>%
  ggplot(aes(x=MDS2, y=MDS3))+
  theme_gray(base_size=12)+
  geom_jitter(data=MDSdf, aes(x=x2,y=x3,col=species), size=2, stroke=1.5,width=0.1, height=0.1, alpha=0.5)+
  geom_convexhull(data=MDSdf, aes(x=x2,y=x3, fill=species), alpha=0.3)+
  scale_fill_manual(values=as.character(levels(plotMDSMRC2$colors)))+
  scale_color_manual(values=as.character(levels(plotMDSMRC2$colors)))+
  geom_segment(aes(xend=0, yend=0))+
  geom_text(aes(label=name))+
  theme_bw(base_size=14)


# without weird points
weirdvec = which((MDSdf$species=="Buffalo" & MDSdf$x1<0)|(MDSdf$species == "Hippo" & MDSdf$x2 <2) )
weirdvec

psitecent %>%
  ggplot(aes(x=MDS2, y=MDS3))+
  theme_gray(base_size=12)+
  geom_jitter(data=MDSdf[-weirdvec,], aes(x=x2,y=x3,col=species), size=2, stroke=1.5,width=0.1, height=0.1, alpha=0.5)+
  geom_convexhull(data=MDSdf[-weirdvec,], aes(x=x2,y=x3, fill=species), alpha=0.3)+
  scale_fill_manual(values=as.character(levels(plotMDSMRC2$colors)))+
  scale_color_manual(values=as.character(levels(plotMDSMRC2$colors)))+
  geom_segment(aes(xend=0, yend=0))+
  geom_text(aes(label=name))+
  theme_bw(base_size=14)

psitecent %>%
  ggplot(aes(x=MDS1, y=MDS2))+
  theme_gray(base_size=12)+
  geom_jitter(data=MDSdf[-weirdvec,], aes(x=x1,y=x2,col=species), size=2, stroke=1.5,width=0.1, height=0.1, alpha=0.5)+
  geom_convexhull(data=MDSdf[-weirdvec,], aes(x=x1,y=x2, fill=species), alpha=0.3)+
  scale_fill_manual(values=as.character(levels(plotMDSMRC2$colors)))+
  scale_color_manual(values=as.character(levels(plotMDSMRC2$colors)))+
  geom_segment(aes(xend=0, yend=0))+
  geom_text(aes(label=name))+
  theme_bw(base_size=14)


MDSdf[weirdvec,]
```

### Repeat for longitudinal data

```{r}


motu_wts = as.data.frame(ordLONG$species)
motu_wts$MOTU =row.names(motu_wts)
names(motu_wts)
merge(motu_wts, MOTU[,c(20:22,24)])
motu_wts = unique(left_join(motu_wts, MOTU[,c(20:22,24)]))
motu_wts$name= motu_wts$name2
#View(motu_wts)
motu_wts$name = as.factor(motu_wts$name)


levels(motu_wts$name) # name and name 2 are off

levels(motu_wts$name) = c("Ancyclostomatidae","Ancyclostomatidae","Ancyclostomatidae","C.elegans","Cloacinidae","Cooperiidae", "Strongylidae","Strongylidae","Haemonchidae","Trichostrongylidae","Ancyclostomatidae","Chabertiidae","Chabertiidae","Protostrongylidae","Strongylida (order)","Strongylidae", "Strongyloidea (super family)","Strongylidae","Haemonchidae","Trichostrongylidae","Trichostrongylidae","Trichostrongyloidea (super family)","Trichostrongylidae","Trichostrongylidae", "Trichostrongylidae", "Protostrongylidae")

motu_wts %>% gather(MDS1:MDS3, key=Axis, value=Score)%>%
ggplot(aes(x=Axis, y=Score))+
  geom_point(aes(group=name,col=name), position=position_dodge(width=0.75))+
  theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0))

# task -- find centroid for each parasite family.

psitecent = motu_wts %>% group_by(name)%>% summarize_at(vars(MDS1:MDS3), funs(mean(.,na.rm=T)))

plotMDSLONG = left_join(plotMDSLONG, colorsp, by=c("Species"="Host"))

plotMDSLONG$Species = factor(plotMDSLONG$Species, levels=levels(reorder(RRA2$Species,-RRA2$richness)))
levels(as.factor(plotMDSLONG$Species))
levels(plotMDSLONG$colors)= c("dodgerblue", "green3","magenta3","cadetblue1","maroon1",  "dodgerblue3","maroon3","deepskyblue3", "maroon", "gold3", "dodgerblue4","green4","deepskyblue", "gold", "lightsteelblue3", "greenyellow",  "darkorchid4","darkorchid1", "lightskyblue2")

ggplot(plotMDSLONG, aes(x=Species, y=x1))+
  geom_point(color=as.character(plotMDSLONG$colors))


psitecent %>%
  ggplot(aes(x=MDS1, y=MDS2))+
  #geom_convexhull(aes(fill=name),alpha=0.3)+
  theme_gray(base_size=12)+
  #scale_color_sjplot(palette="ipsum")+
  #scale_fill_sjplot(palette="ipsum")+
  geom_jitter(data=plotMDSLONG, aes(x=x1,y=x2,col=Species), size=2, stroke=1.5,width=0.1, height=0.1, alpha=0.5)+
  geom_convexhull(data=plotMDSLONG, aes(x=x1,y=x2, fill=Species,linetype=Location), alpha=0.3)+
  scale_fill_manual(values=c("green3","maroon3","dodgerblue3","green4","cadetblue1","darkorchid1"))+
  scale_color_manual(values=c("green3","maroon3","dodgerblue3","green4","cadetblue1","darkorchid1"))+
  geom_segment(aes(xend=0, yend=0))+
  geom_text(aes(label=name))+
  guides(linetype=F)+
  theme_bw(base_size=14)
  #scale_color_sjplot(palette="ipsum")+
  #scale_fill_sjplot(palette="ipsum")


# without weird buffalo
weirdlong = which(plotMDSLONG$Species == "Buffalo" & plotMDSLONG$x2 < -2 | plotMDSLONG$Species =="Kudu" & plotMDSLONG$x1 < -2.5)


psitecent %>%
  ggplot(aes(x=MDS1, y=MDS2))+
  theme_gray(base_size=12)+
  geom_jitter(data=plotMDSLONG[-weirdlong,], aes(x=x1,y=x2,col=Species, shape=Location), size=2, stroke=1.5,width=0.1, height=0.1, alpha=0.5)+
  geom_convexhull(data=plotMDSLONG[-weirdlong,], aes(x=x1,y=x2, fill=Species,linetype=Location), alpha=0.3)+
  scale_fill_manual(values=c("green3","maroon3","dodgerblue3","green4","cadetblue1","darkorchid1"))+
  scale_color_manual(values=c("green3","maroon3","dodgerblue3","green4","cadetblue1","darkorchid1"))+
  scale_shape_manual(values=1:9)+
  geom_segment(aes(xend=0, yend=0))+
  geom_text(aes(label=name))+
  guides(linetype=F)+
  theme_bw(base_size=14)

```


## 7. Creating a network
### 7.1 Setup

```{r}
# Excluding the weird hippo
avgRRA = RRA_present[-which(RRA_present$Sample_ID=="MRC_17_HIP_93"),] %>% group_by(Species) %>% summarize_at(vars(MOTU_1:MOTU_519), funs(mean))

samp_mat_1 = avgRRA


hist(colSums(samp_mat_1[,-1])) # most are extremely low avg RRA
levelplot(as.matrix(samp_mat_1[,-1]),  col.regions = gray(0:100/100))

# Exclude MOTUs that are very rare
samp_mat_2=samp_mat_1[,-(which(colSums(samp_mat_1[,-1])<0.001))]

```

### 7.2 Plotting
```{r}

# need to put all in one column
mat_1_long <- gather(samp_mat_2, key=OTU, value=RRA, -Species)
head(mat_1_long)

# remove zeros
mat_1_long <- mat_1_long[-which(mat_1_long$RRA == 0),]

g <- graph.data.frame(mat_1_long, directed=FALSE)

V(g)$type <- bipartite_mapping(g)$type 
V(g)$color <- ifelse(V(g)$type, "lightblue", "salmon")
V(g)$shape <- ifelse(V(g)$type, "square", "circle")
E(g)$color <- "lightgray"
V(g)$frame.color <-  "white"
V(g)$label.dist <-0.5


plot(g, vertex.label.cex = 0.8, vertex.size=ifelse(V(g)$type,1,7), vertex.label.color = ifelse(V(g)$type,"gray","black"))



# now do this for individuals
#imat_1_long <- gather(RRA_MP_present2[,12:82], key=OTU, value=RRA, -Species)
#imat_1_long <- imat_1_long[-which(imat_1_long$Species %in% c("Primer", "POS", "NEG"))]
#imat_1_long <- imat_1_long[-which(imat_1_long$RRA == 0),]
#imat_1_long <- imat_1_long[-which(is.na(imat_1_long$RRA)),]
#dim(imat_1_long)
#imat_1_long


```

### 7.3 Labeling MOTUs by parasite group
Note that many of these IDs might be mistakes or quite different from the true species identity.

```{r}

# takes a hot second to read in because it's massive
motuinfo = read.delim("nematodes_2018_assembled_setid_filterE2_nl_uniq_c10_ann_assignEMBL_nematoda_sumatra_sumaclust_assignedR134.tab")

#head(motuinfo)
#motuinfo$sequence

motu98 = data.frame(OTU = paste("MOTU_",motuinfo$cluster_sumatra98,sep=""), rank=motuinfo$rank_ok, sp=motuinfo$scientific_name_ok)

head(motu98)
motu98 %>% group_by(OTU) %>% summarize(n())


head(mat_1_long)
head(motu98)

test = left_join(mat_1_long, motu98)

# takes a second
test=unique(test)

# which MOTUs have different sp names
getuni = test %>% group_by(OTU,rank,sp) %>% summarize(n())
getuni[order(getuni[duplicated(getuni$OTU),]$OTU),]

test[which(test$sp=="Cloacinidae"),]
# very weird that this is the closest

# manually get the taxon
test[which(test$OTU=="MOTU_102"),]$sp <- rep("Strongyloidea",2)
test[which(test$OTU=="MOTU_11"),]$sp <- rep("Strongylida",6)
test[which(test$OTU=="MOTU_112"),]$sp <- rep("Trichostrongylidae",2)
test[which(test$OTU=="MOTU_12"),]$sp <- rep("Strongylida",4)
test[which(test$OTU=="MOTU_123"),]$sp <- rep("Trichostrongylus",4)
test[which(test$OTU=="MOTU_126"),]$sp <- rep("Trichostrongylus",2)
test[which(test$OTU=="MOTU_14"),]$sp <- rep("Trichostrongylus vitrinus",10)

test[which(test$OTU=="MOTU_2"),]$sp <- rep("Strongylida",4)# or 6
test[which(test$OTU=="MOTU_18"),]$sp <- rep("Strongylida",2)
test[which(test$OTU=="MOTU_23"),]$sp <- rep("Strongyloidea",2)
test[which(test$OTU=="MOTU_28"),]$sp <- rep("Strongylida",2) #or4
test[which(test$OTU=="MOTU_30"),]$sp <- rep("Oesophagostomum",8)#8

test[which(test$OTU=="MOTU_26"),]$sp <- rep("Strongylida",2)
test[which(test$OTU=="MOTU_40"),]$sp <- rep("Trichostrongylidae",8)
test[which(test$OTU=="MOTU_41"),]$sp <- rep("Strongylida",2)
test[which(test$OTU=="MOTU_42"),]$sp <- rep("Trichostrongylus",8)#10
test[which(test$OTU=="MOTU_51"),]$sp <- rep("Strongylida",2)#4
#test[which(test$OTU=="MOTU_529"),]$sp <- rep("Oesophagostomum",2)
test[which(test$OTU=="MOTU_55"),]$sp <- rep("Strongyloidea",2)
test[which(test$OTU=="MOTU_64"),]$sp <- rep("Trichostrongylidae",2)
test[which(test$OTU=="MOTU_68"),]$sp <- rep("Oesophagostomum",6)
test[which(test$OTU=="MOTU_8"),]$sp <- rep("Strongylida",2)
test[which(test$OTU=="MOTU_85"),]$sp <- rep("Trichostrongylidae",2)
test[which(test$OTU=="MOTU_90"),]$sp <- rep("Strongylida",2)
test[which(test$OTU=="MOTU_60"),]$sp <- rep("Strongylida",2)
test[which(test$OTU=="MOTU_74"),]$sp <- rep("Trichostrongylus vitrinus",4)
#test[which(test$OTU=="MOTU_98"),]$sp <- rep("Strongyloidea",2)
#test[which(test$OTU=="MOTU_122"),]$sp <- rep("Trichostrongylus",3)
test[which(test$OTU=="MOTU_136"),]$sp <- rep("Strongylida",2)
test[which(test$OTU=="MOTU_205"),]$sp <- rep("Necator",2)
test[which(test$OTU=="MOTU_59"),]$sp <- rep("Strongylida",2)
test[which(test$OTU=="MOTU_28"),]$sp <- rep("Strongylida",2)
test[which(test$OTU=="MOTU_75"),]$sp <- rep("Trichostrongylidae",2)


test %>% group_by(OTU,rank,sp) %>% summarize(n())
test2 = (unique(test[,-4]))

test2 %>% group_by(OTU,sp) %>% summarize(n())

dim(test2) 
```

```{r}
head(test2)
# need to make a column with sp and motu
test2[order(test2$sp),]

test2$sp2 = paste(test2$sp,str_remove(test2$OTU, "MOTU_"),sep="_")


# simplify by removing any connections when the avg RRA less than 0.1%???
length(which(test2$RRA <0.001))
test3 = test2[-which(test2$RRA <0.001),]


# also remove the strange MOTU 37 results
head(test3)
test3 = test3[-which(test3$OTU %in% c("MOTU_37","MOTU_96") & test3$Species %in% c("Plains zebra","Grevys zebra","Kaia","Donkey")),]


new_mat = test3[,c(1,5,3)]
g <- graph.data.frame(new_mat, directed=FALSE)

V(g)$type <- bipartite_mapping(g)$type 
V(g)$color <- ifelse(V(g)$type, "lightblue", "salmon")
V(g)$shape <- ifelse(V(g)$type, "square", "circle")
E(g)$color <- "lightgray"
V(g)$frame.color <-  "white"
V(g)$label.dist <-0.5


plot(g, vertex.label.cex = 0.8, vertex.size=ifelse(V(g)$type,3,8), vertex.label.color = ifelse(V(g)$type,"gray","black"))


```


### 7.3 Create clusters based on Jaccard distance

```{r}


test4 = test3[,-c(4,5)] %>% spread(key=OTU, value=RRA, fill=0)
dim(test4)

allRRA = unlist(test4[,-1])
hist(log(allRRA[allRRA>0]))

#test4[test4$Species=="PLA",]$MOTU_37 <- 0

test5 = as.matrix(test4[,-1])
row.names(test5)=test4$Species

levelplot(as.matrix(test5))

ggplot(test3[test3$Species %in% c("Plains zebra","Grevys zebra"),], aes(x=OTU,y=RRA))+
  geom_col(aes(fill=Species))


distmat = distance(test5, method="jaccard")
row.names(distmat)=row.names(test5)
colnames(distmat)=row.names(test5)


Heatmap(distmat, col=c("lightblue","black"), column_km=5, row_km_repeats=10000)

```

### 7.4 Network of only impala, cattle, and elephant

This was used as a demonstration in a talk

```{r}

head(test3)
small_mat = test3[test3$Species %in% c("Impala","Cattle","Elephant"),c(1,5,3)]

g <- graph.data.frame(small_mat, directed=FALSE)

V(g)$type <- bipartite_mapping(g)$type 
V(g)$shape <- ifelse(V(g)$type, "circle", "circle")
E(g)$color <- "gray"
E(g)$weights = small_mat$RRA

V(g)$frame.color <-  "black"
V(g)$label.dist <-1.5



vertex_attr(g)$color<-rep("#FF9C0B", length(V(g))) #coloring all the nodes orange
#Selecting everyone that is "TRUE" in $type to change it to green (take a look at the help of the grep function in case you don't know it).
vertex_attr(g)$color[grep(pattern = "FALSE", vertex_attr(g)$type)]<-"#1C912C"
vertex_attr(g)$color # checking the color vector

vertex_attr(g)$frame.color<-rep("#FF9C0B", length(V(g))) #coloring all the nodes orange
#Selecting everyone that is "TRUE" in $type to change it to green (take a look at the help of the grep function in case you don't know it).
vertex_attr(g)$frame.color[grep(pattern = "FALSE", vertex_attr(g)$type)]<-"#1C912C"
vertex_attr(g)$frame.color # checking the color vector

plot(g, vertex.label.cex = ifelse(V(g)$type,0.9,1.5), vertex.size=ifelse(V(g)$type,10,16), vertex.label.color = ifelse(V(g)$type,"gray50","black"), edge.curved=0.3,  edge.width=20*sqrt(edge_attr(g)$weight))


```


## 8. Investigating possible zebra anomalies


```{r}

zebt3 = test3[which(test3$Species %in% c("Plains zebra","Grevys zebra","Kaia","Donkey")),]
ggplot(zebt3,aes(x=sp, y=RRA))+
  geom_boxplot(aes(col=Species))

#96,37 MOTUs in zeb

zebt3[which(zebt3$OTU %in% c("MOTU_96","MOTU_37")),]
```

```{r}

# going back to RRA2 to check out the individuals

#head(RRA2)
MOTU37 = ggplot(RRA2, aes(x=Species,y=MOTU_37))+
  geom_boxplot(aes())+
  theme(axis.text.x=element_text(angle=90))
MOTU96 = ggplot(RRA2, aes(x=Species,y=MOTU_96))+
  geom_boxplot(aes())+
  theme(axis.text.x=element_text(angle=90))

gridExtra::grid.arrange(MOTU37,MOTU96)


# 37 is more of a problem than 96

odd_samples = RRA2[which(RRA2$MOTU_37 >0 & RRA2$Species%in% c("Grevys zebra","Plains zebra","Kaia","Donkey")),]$Sample_ID

odd_samples

badones = OTUtab[OTUtab$X %in% odd_samples,]


badlong = badones %>% gather(key=MOTU, value=reads,-X)
badlong$reads = as.numeric(badlong$reads)
badlong=badlong[-which(is.na(badlong$reads)),]

ggplot(badlong,aes(x=X, y=as.numeric(reads)))+geom_col(fill="gray", alpha=0.5)+
  geom_col(data=badlong[badlong$MOTU=="MOTU_37",], aes(x=X, y=reads), fill="red",alpha=0.5)+
  theme(axis.text.x=element_text(angle=90))+
  labs(x="Sample",y="Total Reads")

```

One reassuring thing -- these samples have a strong signature for other OTUs that are indicative of zebras.

I have to assume that this is chance weird occurence or contamination of some sort. Another option is that they consumed these nems but were not infected by them. Still, the result is a bit weird and I wish that they could be filtered....


