---
title: "2021_Metabarcoding_Analyses"
author: "Georgia Titcomb"
date: "January 31, 2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(vegan)
library(GUniFrac)
library(sjPlot)
library(readxl)
library(ggConvexHull)
library(igraph)
library(phylosignal)
library(adephylo)
library(ape)
library(phylobase)
library(philentropy)
library(ComplexHeatmap)
library(lmerTest)
library(emmeans)
library(glmmTMB)
library(DHARMa)
library(MuMIn)
library(picante)
library(NetworkExtinction)
library(msa)
library(adegenet)
library(phyloseq)
library(QsRutils)
library(phangorn)
library(seqinr)
library(bipartite)
library(MCMCglmm)
library(binom)
library(here)


```

### Prepare RRA data

## 1. Import data and finish filtering

```{r}

source("prepare_RRA_table.R")
source("functions.R")
```


## 2. Initial richness comparison

### 2.1 Calculate MOTU richness

```{r}

# Add MOTU richness for each sample (after rarifying to same reads)
RRA2$richness = specnumber(RRA2[,30:195])

# Does individual richness vary significantly by species?

#non zeros
RRA_present = RRA2[which(RRA2$richness>0),]
dim(RRA_present)

# RRA2 is all data
# RRA_present is richness > 0 

# set animal colors based on alphabet
animal_colors=c("darkorchid4","goldenrod1","blueviolet", "deepskyblue3", "hotpink",  "dodgerblue","green3",  "goldenrod","dodgerblue3", "maroon1",  "deepskyblue2", "greenyellow", "dodgerblue4", "lightskyblue","lightskyblue1", "maroon3", "green4", "cyan2")

# plot only MPdata
withzeros = ggplot(RRA2, aes(x=reorder(Species,-richness), y=richness))+
  geom_jitter(width=0.01,height=0.3, alpha=0.5, aes(color=Species))+
  geom_boxplot(aes(fill=Species), color="gray50",outlier.shape=NA)+
  guides(fill=F,col=F)+
  theme_bw()+
  scale_fill_manual(values=animal_colors)+
  scale_color_manual(values=animal_colors)+
  theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0))+
  labs(x="Host Species",y="Rarified OTU richness",title="Individual-level Richness (with 0s)")

nozeros = ggplot(RRA_present, aes(x=reorder(Species,-richness,na.rm=T), y=richness))+
  geom_jitter(width=0.01,height=0.3, alpha=0.5, aes(color=Species))+
  geom_boxplot(aes(fill=Species), color="gray50",outlier.shape=NA)+
  guides(fill=F,col=F)+
  theme_bw(base_size=16)+
  scale_fill_manual(values=animal_colors)+
  scale_color_manual(values=animal_colors)+
  theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0))+
  labs(x="Host Species",y="Richness",title="Richness")

richgraph = gridExtra::grid.arrange(withzeros, nozeros, ncol=2)
#ggsave("g98_plots/rich_graph.png", richgraph, device="png")

### Added
# prevalence graph
# calculate prevalence with correct errorbars
MPdata_prev = RRA2 %>% group_by(Species) %>% summarize_at(vars(overall_prev), funs(n=n(), pos=sum(.)))

MPdata_prev = binom.confint(MPdata_prev$pos, MPdata_prev$n, methods="wilson")%>%cbind(MPdata_prev[,1])

prevalence = ggplot(MPdata_prev, aes(x=reorder(Species, -mean), y=mean))+
  geom_point(aes(color=Species), size=2)+
  geom_errorbar(aes(ymin=lower, ymax=upper, color=Species), size=1.2)+
  guides(col=F)+
  theme_bw(base_size = 16)+
  scale_color_manual(values=animal_colors)+
  theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0))+
  scale_y_continuous(limits=c(0,1))+
  labs(x="Host Species", y="Prevalence", title="Prevalence")

richgraph = gridExtra::grid.arrange(prevalence, nozeros, ncol=2)
#ggsave(here("g98_plots/prev_richgraph.png"),richgraph, width=10, height=5, units="in", dpi=300, device="png")
```


### 2.2 Conduct statistical tests of richness by species

```{r}

# fit model
MPsp = glmmTMB(richness ~ Species, family="poisson", data=RRA_present)
# diagnostics
DHARMa::testResiduals(simulateResiduals(MPsp))
car::Anova(MPsp)

# Table 1 data
# multiple comparisons
emmp = emmeans(MPsp, ~Species, data=RRA_present, type="response")
pairs(emmp, adjust="fdr") %>% as.data.frame() %>% filter(p.value<0.05)

# results
tab_model(MPsp, transform = NULL, show.stat=T, show.se = T)


# Test prevalence
MPprev = glmmTMB(overall_prev ~ Species, data=RRA2, family="binomial")
tab_model(MPprev, transform = NULL) # R2 is 18.1% 
car::Anova(MPprev)

# Table 1 data
emmp_prev = emmeans(MPprev, ~Species, type="response")
pairs(emmp_prev, adjust="fdr")%>%as.data.frame()%>%filter(p.value<0.05)

```

Richness and prevalence clearly vary by species. 60% and 18% of variation in richness is explained by species. Richness: $\chi^2_{18} = 224.2, p < 0.001$; Prevalence: $\chi^2_{18} = 64.89, p < 0.001$.

### 2.3 Richness and prevalence and host characteristics

```{r}
# Richness 

# Fit a model with a species term (includes zeros - quasipoisson)
RRA2$phylo=RRA2$MSW93_Binomial

# Exclude camel and hippo as they are pseudoruminants
RRA2_noCH = RRA2[-which(RRA2$Species %in% c("Camel","Hippo")),]
dim(RRA2_noCH)

# Make another positives-only version for richness
RRA2_noCH_pos = RRA2_noCH[which(RRA2_noCH$richness>0),]
dim(RRA2_noCH_pos)

```


```{r}
# format tree
tree = ape::read.tree(here("data/mammaltree.txt"))
inv.phylo = inverseA(tree, nodes="TIPS", scale=T) # decide all or tips

# set prior for poisson
prior_pois = list(G=list(G1=list(V=1,nu=0.02),G2=list(V=1,nu=0.02)), R=list(V=1,nu=0.02))

# format data correctly
MPrichinfoMCMC = as.data.frame(RRA2_noCH_pos)
MPrichinfoMCMC$animal = MPrichinfoMCMC$Sample_ID
MPrichinfoMCMC$phylo = MPrichinfoMCMC$MSW93_Binomial
MPrichinfoMCMC = as.data.frame(MPrichinfoMCMC)

# Fit model
set.seed(123)
model_rich = MCMCglmm(richness ~ Diet + Feeding_Type +log(Body_Mass_kg)+ log(SocialGrpSize)+ log(HomeRange_km2), random= ~ phylo + MSW93_Binomial, family = "poisson", ginverse=list(phylo=inv.phylo$Ainv), prior=prior_pois, data=MPrichinfoMCMC, nitt=500000, burnin=10000, thin=250, verbose=F, pl=T)

# check autocorrelation
autocorr(model_rich$Sol)
autocorr(model_rich$VCV) 

sumrich = make_MCMC_summary(model_rich)
flextable:::knit_print.flextable(sumrich)
```

```{r}

# model selection
MCMCglmm.updateable = updateable(MCMCglmm)
set.seed(123)
model_richt = MCMCglmm.updateable(richness ~ Diet + Feeding_Type + log(Body_Mass_kg) + log(SocialGrpSize) + log(HomeRange_km2), random= ~phylo + MSW93_Binomial, family = "poisson", ginverse=list(phylo=inv.phylo$Ainv), prior=prior_pois, data=MPrichinfoMCMC, nitt=10000, burnin=5000, thin=50, verbose=F, pl=T)
drt = dredge(model_richt, rank="DIC")
drt

# Best model
set.seed(0)
model_simple2 = MCMCglmm(richness ~Diet, random= ~phylo + MSW93_Binomial, family = "poisson", ginverse=list(phylo=inv.phylo$Ainv), prior=prior_pois, data=MPrichinfoMCMC, nitt=500000, burnin=10000, thin=250, verbose=F, pl=T)

# assess model
sumrich_red = make_MCMC_summary(model_simple2)
flextable:::knit_print.flextable(sumrich_red)

# sanity check
rmod = glmmTMB(richness ~ Diet + (1|MSW93_Binomial), family="poisson", data=MPrichinfoMCMC)
summary(rmod)

rmodem = emmeans(rmod, ~Diet, type="response")
rmodem

```

### 2.4 Prevalence and host characteristics

```{r}

#################################
###### Prevalence
head(RRA2_noCH)

# advice from https://r-sig-mixed-models.r-project.narkive.com/I9vUTpvM/r-sig-me-rare-binary-outcome-mcmcglmm-and-priors
prior_binom = list(R = list(V = 1, fix = 1), B = list(mu = c(rep(0,7)), V = diag(7)*(4.3+pi^2/3)),
G = list(G1 =list(V = 1, nu = 0.002, alpha.mu = 0, alpha.V = 1000),G1 =list(V = 1, nu = 0.002, alpha.mu = 0, alpha.V = 1000)))

# format data correctly
MPrichinfoMCMC = as.data.frame(RRA2_noCH)
MPrichinfoMCMC$animal = MPrichinfoMCMC$Sample_ID
MPrichinfoMCMC$phylo = MPrichinfoMCMC$MSW93_Binomial
MPrichinfoMCMC = as.data.frame(MPrichinfoMCMC)
dim(MPrichinfoMCMC)


# fast version
MCMCglmm.updateable = updateable(MCMCglmm)
set.seed(0)
modt = MCMCglmm.updateable(fixed = overall_prev ~ Diet + Feeding_Type + log(Body_Mass_kg) + log(SocialGrpSize) + log(HomeRange_km2),
                           random = ~phylo + MSW93_Binomial,
                           ginverse = list(phylo = inv.phylo$Ainv),
                           data = MPrichinfoMCMC,
                           family = "categorical",
                           prior = prior_binom,
                           nitt = 500000, burnin = 10000, thin = 250,
                           verbose = F, pl = T) 

sumbin = make_MCMC_summary(modt)
flextable:::knit_print.flextable(sumbin)

```

```{r}
# body mass and diet are potential coefs
prior_binom1 = list(R = list(V = 1, fix = 1),
                    B = list(mu = c(rep(0,4)), V = diag(4)*(4.3+pi^2/4)),
                    G = list(G1 =list(V = 1, nu = 0.002, alpha.mu = 0, alpha.V = 1000),
                        G1 =list(V = 1, nu = 0.002, alpha.mu = 0, alpha.V = 1000)))
set.seed(0)
mod0.1 = MCMCglmm.updateable(fixed = overall_prev ~ Diet + log(Body_Mass_kg),
                           random = ~phylo + MSW93_Binomial,
                           ginverse = list(phylo = inv.phylo$Ainv),
                           data = MPrichinfoMCMC,
                           family = "categorical",
                           prior = prior_binom1,
                           nitt = 10000, burnin = 5000, thin = 50,
                           verbose = F, pl = T) 
summary(mod0.1)

# compare DIC value to model without body mass

prior_binom2 = list(R = list(V = 1, fix = 1),
                    B = list(mu = c(rep(0,3)), V = diag(3)*(4.3+pi^2/3)),
                    G = list(G1 =list(V = 1, nu = 0.002, alpha.mu = 0, alpha.V = 1000),
                        G1 =list(V = 1, nu = 0.002, alpha.mu = 0, alpha.V = 1000)))
# reduced
set.seed(0)
mod0.2 = MCMCglmm(fixed = overall_prev ~ Diet,
                  random = ~ phylo + MSW93_Binomial,
                  data = MPrichinfoMCMC,
                  family = "categorical",
                  prior = prior_binom2,
                  ginverse=list(phylo=inv.phylo$Ainv),
                  nitt=500000, burnin=10000, thin=250,
                  verbose=F, pl=T) 

# check autocorrelation
autocorr(mod0.2$Sol)
autocorr(mod0.2$VCV) 

sumbin_red = make_MCMC_summary(mod0.2)
flextable:::knit_print.flextable(sumbin_red)

# compare estimages to GLMM for sanity check -- they are not too far off
dim(MPrichinfoMCMC) # should be 495

model = glmmTMB(overall_prev ~ Diet + Feeding_Type + log(Body_Mass_kg) + log(SocialGrpSize)+ log(HomeRange_km2)+ (1|MSW93_Binomial), family="binomial", data=MPrichinfoMCMC)
summary(model)
model = glmmTMB(overall_prev ~ Diet + (1|MSW93_Binomial), family="binomial", data=MPrichinfoMCMC)
summary(model)

emglmm = emmeans(model, ~ Diet, type="response")
pairs(emglmm)
```

### 2.5 Plots

```{r}
## Raw data and plotting
MPprev = RRA2_noCH %>% group_by(Diet, Feeding_Type) %>% summarize_at(vars(overall_prev), funs(n=n(), pos=sum(.)))

#library(binom)
MPprev = binom.confint(MPprev$pos, MPprev$n, methods="wilson")%>%cbind(MPprev[,1:2])

MPprev$Diet=factor(MPprev$Diet, levels=c("G","M","B"))
MPdata_prev2 = unique(left_join(as.data.frame(MPdata_prev), dplyr::select(RRA2, Species, Diet, Feeding_Type), by="Species"))
levels(as.factor(MPdata_prev2$Feeding_Type))
levels(as.factor(MPdata_prev2$Species))


MPdata_prev2 = MPdata_prev2[-which(MPdata_prev2$Species %in% c("Camel","Hippo","Hybrid zebra")),]

MPprev$Diet = factor(MPprev$Diet, levels=c("B","M", "G"))
MPdata_prev2$Diet = factor(MPdata_prev2$Diet, levels=c("B","M","G"))

prevplotgp = ggplot(MPprev, aes(x=Diet, y=mean))+
  geom_errorbar(data=MPdata_prev2, aes(x=Diet, y=mean, ymin=lower, ymax=upper, color=Species), size=0.9, position=position_dodge(width=0.5), width=0.25, alpha=0.6)+
  geom_point( position=position_dodge(width=0.75), size=2, shape=1)+
  geom_errorbar(aes(ymin=lower, ymax=upper),position=position_dodge(width=0.75), size=1.2, col="gray50")+
  theme_bw(base_size=16)+
  labs(x="Diet", y="Prevalence", title="Prevalence")+
  scale_color_manual(values=c(animal_colors[c(1,3:7)], animal_colors[c(8:11,13)],animal_colors[14:18]))+
  scale_y_continuous(limits=c(0,1))+
  facet_wrap(~Feeding_Type, scales="free_x")+guides(col=F)
  

# same for richness
RRA2_noCH_pos$Diet=factor(RRA2_noCH_pos$Diet, levels=c("B","M","G"))

richplotgp = ggplot(RRA2_noCH_pos, aes(x=Diet, y=richness))+
  geom_boxplot(data=RRA2_noCH_pos, aes(x=Diet, y=richness,color=Species), size=0.7, position=position_dodge(width=0.5), width=0.25, alpha=0.6)+
  geom_boxplot(col="gray50", fill=NA, size=1)+
  theme_bw(base_size=16)+
  labs(x="Diet", y="Richness", title="Richness (no 0s)")+
  scale_color_manual(values=c(animal_colors[c(1,3:7)], animal_colors[c(8:11,13)],animal_colors[14:18]))+
  scale_y_continuous(limits=c(0,18))+
  facet_wrap(~Feeding_Type, scales="free_x")+guides(col=F)

gpplots = gridExtra::grid.arrange(prevplotgp, richplotgp, ncol=2)

#ggsave(here("g98_plots/gpplots.png"),gpplots, device="png", dpi=300, width=10, height=5, units="in")
```



## 3. Comparing host range among MOTUs

```{r}
# ADD up RRA for each species?
MOTU = RRA2 %>% group_by(Species) %>%
  summarize_at(vars(MOTU_1:MOTU_557),funs(mean(.))) %>%
  as.data.frame()%>%
  .[,-1] %>%
  t()%>%
  as.data.frame()

names(MOTU) = levels(as.factor(RRA2$Species))

# Binarize
MOTU = MOTU %>% mutate_at(vars(Buffalo:Waterbuck), funs(ifelse(.>0,1,0)))
MOTU$Hosts = rowSums(MOTU)
MOTU$MOTU = names(dplyr::select(RRA2, MOTU_1:MOTU_557))

ggplot(MOTU, aes(x=Hosts, y=reorder(MOTU,Hosts)))+
geom_jitter(width=0.1,height=0.1,alpha=0.5)+
scale_x_continuous(breaks=c(1,2,3,4,5,6,7,8,9,10))

hstct = data.frame(host = names(colSums(MOTU[,-c(19,20)])), count=colSums(MOTU[,-c(19,20)]))
ggplot(hstct, aes(x=count, y=reorder(host,count)))+
  geom_jitter()


MOTU_long = MOTU %>% pivot_longer(cols=Buffalo:Waterbuck, names_to="Host", values_to="link")
MOTU_long = left_join(MOTU_long, RRA2[,c(3,26)], by=c("Host"="Species"))
head(MOTU_long)

```


### 3.1 Add taxonomic information

```{r}
# bring in MOTU information
motuinfo = read.csv(here("data/NEW_TAXA_sequences_80.csv"))

motu98 = data.frame(cluster = (motuinfo$cluster_sumatra98),
                    family_j = motuinfo$family_name_ok,
                    name_j = motuinfo$scientific_name_ok,
                    order_g = motuinfo$Order,
                    family_g = motuinfo$Genus,
                    name_g = motuinfo$unknown)

motu98 = unique(motu98)

motu98$MOTU = paste("MOTU_", motu98$cluster, sep="")


# rerun MOTU code before running this
MOTU = left_join(MOTU, motu98, by="MOTU")

# create a grouping variable to color by parasite family
MOTU$name2 = MOTU$family_g
MOTU$name2 = as.character(MOTU$name2)
MOTU2 = dplyr::select(MOTU, MOTU, name2, Hosts) %>%
  unique()

# remove duplicates and missing
MOTU2 = MOTU2[order(MOTU2$name2),]
MOTU2 = MOTU2[-which(is.na(MOTU2$name2) & duplicated(MOTU2$MOTU)),]

# Some IDs are missing, so we can use the previous ID
missing = MOTU2[which(is.na(MOTU2$name2)),]$MOTU
MOTU[which(MOTU$MOTU %in% missing),] %>%
  dplyr::select(MOTU, name_j) %>%
  unique()

# this has to be done manually
dfmiss = data.frame(MOTU=missing, newname = c("Strongyloidea",
                                              "Trichostrongyloidae",
                                              "Strongyloidea",
                                              "Strongyloidea",
                                              "Trichostrongylidae",
                                              "Strongyloidea",
                                              "Strongyloidea",
                                              "Trichostrongylidae",
                                              "Strongyloidea",
                                              "Strongyloidea",
                                              "Strongyloidea",
                                              "Strongyloidea",
                                              "Strongyloidea",
                                              "Strongyloidea",
                                              "Ancyclostomatidae",
                                              "Ancyclostomatidae",
                                              "Ancyclostomatidae",
                                              "Trichostrongylidae",
                                              "Strongylida"))

MOTU2[which(is.na(MOTU2$name2)),]$name2 = dfmiss$newname

# Drop name_j and regroup
MOTU2 = dplyr::select(MOTU2, MOTU, name2, Hosts)%>%unique()

# check levels
levels(as.factor(MOTU2$name2))

# clean up levels
MOTU2$name2 = as.factor(MOTU2$name2)
levels(MOTU2$name2) = c("Ancyclostomatidae (hookworms)", "Ancyclostomatidae (hookworms)", "Chabertiidae (nodular worms)","Cooperiidae (cooperia)", "Haemonchidae (barber pole worms)", "Protostrongylidae (lungworms)", "Strongylida (order)", "Strongylidae (strongyles)", "Strongyloidea (superfamily)", "Trichostrongylidae (trichostrongyles)", "Trichostrongylidae (trichostrongyles)")
levels(MOTU2$name2)
# reorder levels
MOTU2$name2 = factor(MOTU2$name2, levels=c("Ancyclostomatidae (hookworms)", "Chabertiidae (nodular worms)", "Cooperiidae (cooperia)", "Haemonchidae (barber pole worms)", "Trichostrongylidae (trichostrongyles)", "Protostrongylidae (lungworms)", "Strongylidae (strongyles)", "Strongyloidea (superfamily)", "Strongylida (order)"))


# exclude 0s
MOTU2 = MOTU2[-which(MOTU2$Hosts==0),]%>% unique()

# now reorder MOTUs based on Hosts AND name2
MOTU2 = MOTU2[order(MOTU2$name2, MOTU2$Hosts),] # this should be the correct order
MOTUorder = unique(MOTU2$MOTU)
MOTU2$MOTU = factor(MOTU2$MOTU, levels=MOTUorder)

psite_distribution = ggplot(MOTU2, aes(x=Hosts, y=MOTU))+
  geom_point(aes(col=name2), size=1.6)+
  geom_line(aes(col=name2, group=name2), size=1)+
  scale_x_continuous(breaks=c(1,2,3,4,5,6,7,8,9,10))+
  scale_y_discrete(labels=NULL)+
  #theme_classic(base_size = 14)+
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        panel.border = element_blank(),
        axis.line=element_line(color="gray"),
        text= element_text(size=14),
        axis.text.x = element_text(size=14))+
  guides(col=F)+
  facet_wrap(~name2, nrow=7, scales="free_y")+
  labs(x="Number of Host Species", y="mOTU ID")
psite_distribution

#ggsave(here("g98_plots/psite_host_distribution2.png"), psite_distribution, device="png", width=3, height=12)
```

```{r}

# now reorder MOTUs based on Hosts AND name2
MOTU2 = MOTU2[order(MOTU2$Hosts, MOTU2$name2),] # this should be the correct order
MOTUorder = unique(MOTU2$MOTU)
MOTU2$MOTU = factor(MOTU2$MOTU, levels=MOTUorder)

ggplot(MOTU2, aes(x=Hosts, y=MOTU))+
  geom_jitter(width=0.1,height=0.1,alpha=0.5, aes(color=name2), size=2)+
  scale_x_continuous(breaks=c(1,2,3,4,5,6,7,8,9,10))+
  scale_y_discrete(labels=NULL)+
  theme_light(base_size=14)

hist(MOTU2$Hosts[which(MOTU2$Hosts>0)])
mtudisvec = MOTU2$Hosts[which(MOTU2$Hosts>0)]
length(which(mtudisvec==1))/length(mtudisvec) 
length(which(mtudisvec>7))/length(mtudisvec)
MOTU[which(MOTU$Hosts==10),] # MOTU 5

```


## 4. Sequence-sequence relationships

### 4.1 Subset the sequence data to the filtered MOTUs

```{r}
# Streamlined make_fasta script pasted below:
# Constructed using all data

# first get the list of MOTUs of interest
motunames = read.csv(here("data/RRAtab_filt2_g98.csv"))

# trim dataframe (seqs replaced by motuinfo)
seqsmall = motuinfo[,c(1,7,9,18,22,27,28)]
seqsmall$MOTU = paste("MOTU_",seqsmall$cluster_sumatra98, sep="")

# find corresponding sequences
relevant_seqs = seqsmall[match(seqsmall$MOTU,names(motunames)[-1]),]
dim(relevant_seqs)
dim(seqsmall) #64291

# now select the most abundant sequence for each MOTU
topseq = relevant_seqs %>% group_by(MOTU) %>% top_n(1, count)

# add taxa info
topseq = left_join(topseq, motuinfo[,18:28])

# now use only these sequences
# make a subset
myfasta= read.fasta(file = here("data/nematode.fasta"), seqtype = "AA",as.string = TRUE, set.attributes = FALSE)

my_fasta_sub = myfasta[names(myfasta) %in% topseq$id]

# save subset
write.fasta(sequences = my_fasta_sub, names = names(my_fasta_sub), nbchar = 80, file.out = here("data/nem_subset.fasta"))

```

### 4.2 Combine host and sequence data for later plotting

```{r}
# find the animal that is most common for each motu
RRAagg = RRA2 %>%
  group_by(Species) %>%
  summarize_at(vars(MOTU_1:MOTU_557), funs(mean)) %>% 
  ungroup()

# which species has the highest relative frequency for each MOTU?
specs = unlist(apply(RRAagg[,-1], 2, which.max))
specieslink = RRAagg$Species[specs]


# how about if the tip label is the number of species?
RRA2_bin = dplyr::select(RRA2, MOTU_1:MOTU_557) %>%
  mutate_at(vars(MOTU_1:MOTU_557), funs(ifelse(.>0,1,0)))

RRA2_bin$Species = RRA2$Species

# use only positives because high prev species overall will bias
RRA2_bin_present = RRA2_bin[-which(rowSums(RRA2_bin[,-c(167)])==0),]
RRA2_binagg = RRA2_bin_present %>%
  group_by(Species) %>%
  summarize_at(vars(MOTU_1:MOTU_557), funs(mean)) %>% 
  ungroup()

numberhost = RRAagg[,-1] %>%
  mutate_at(vars(MOTU_1:MOTU_557), funs(ifelse(.>0,1,0))) %>%
  colSums()

binnumberhost = RRA2_binagg[,-1] %>%
  mutate_at(vars(MOTU_1:MOTU_557), funs(ifelse(.>0.0,1,0))) %>%
  colSums()

```

### 4.3 Align sequences and create tree
Skip this to avoid recalculating everything

```{r}
source("prepare_tree.R")
```

### 4.4 Plot tree with metadata

```{r}
#Plotting with metadata
treeNJ = read.tree(here("data/g98_treeNJ_K80_gamma_check.tree")) # manually mutated neg vals to pos 

splink = data.frame(species=specieslink, richness = binnumberhost, MOTU=names(RRAagg[-1]))
treeNJ$tip.label = motuinfo$MOTU[match(treeNJ$tip.label,motuinfo$id)]

zeroR = splink %>% 
  filter(richness!=0) %>% 
  dplyr::select(MOTU) %>% 
  unlist()

treeNJ2 = prune_taxa(zeroR, treeNJ)

ttlab = data.frame((topseq$MOTU[which(topseq$MOTU %in% treeNJ2$tip.label)]))
ttlab = topseq[match(treeNJ2$tip.label,topseq$MOTU),]
names(ttlab)[2]="MOTU"
ttlab$MOTU = paste("MOTU",ttlab$MOTU, sep="_")
ttlab = as.data.frame(ttlab)
ttlab = left_join(ttlab[,c(1,2,5,6,7,8)], splink, by=c("MOTU"))

myPal <- colorRampPalette(c("red","yellow","green","blue"))

# phylogeny colored by nem taxonomy

png(here("g98_plots/trees.png"), res=300, width=8, height=8, units="in")
par(mfrow=c(1,2))
plot(treeNJ, show.tip=F, x.lim = 0.85)
tiplabels(ttlab$scientific_name_ok, cex=0.5, col=transp(num2col(ttlab$bid_ok*100, col.pal=myPal), 0.7), frame="n", offset=0, adj=0)
temp = pretty(min(ttlab$bid_ok*100):max(ttlab$bid_ok*100),5)
legend("topright",fill=transp(num2col(temp, col.pal=myPal),.7), leg=temp, ncol=1, cex=0.7, title="best identity score")

# phylogeny colored by core host
plot(treeNJ2, cex=0.6, show.tip=F, align.tip.label=F, x.lim=0.85)
tiplabels(ttlab$species, cex=0.5, frame="n", col=transp(num2col(ttlab$richness, col.pal=myPal), 0.7), offset=0, adj=0)
temp <- c(1,2,3,5,10)
legend("topright", fill=transp(num2col(temp, col.pal=myPal),.7), leg=c(1,2,3,5,10), ncol=1, cex=0.7, title="number of host species")
dev.off()

```

### 4.5 Create phyloseq object and find Unifrac Distances

#### 4.5.i Individual-level (samples)

```{r}
### need to make a phyloseq object
motunames = read.csv(here("data/RRAtab_filt2_g98.csv"))

#treeNJ$tip.label == ttlab$MOTU
treetree = phy_tree(treeNJ2)
# format for phyloseq
treetree = multi2di(treetree)
treetree = (root(treetree,1))

# create otu table (this is all samples)
otuotu = otu_table(motunames[,-1], taxa_are_rows = F)

# make otu table for mpala speccies only
motunames_with_sp = left_join(motunames, RRA2[,1:4])
mtusp = motunames_with_sp[which(motunames_with_sp$Location=="Mpala"),]
otuotu = otu_table(mtusp[,-c(1,168:170)], taxa_are_rows =F)

# combine into phyloseq object. Note that the tree probably contains some branches that are not covered by all samples in Mpala. But my thought is that more information makes a better tree.
comboseq = phyloseq(treetree,otuotu)

# run unifrac
set.seed(0)
ufc = UniFrac(comboseq, normalized=F, weighted =F) # is already normalized
hist(ufc)

# ufc are the unifrac distances -- save this as a file
ufcmat = as.matrix(ufc)
rownames(ufcmat)=mtusp$Sample_ID[which(mtusp$Location=="Mpala")]
colnames(ufcmat)=mtusp$Sample_ID[which(mtusp$Location=="Mpala")]

# mutate out of bounds distances
ufctab = ufcmat %>% as.tibble()

ufctab = ufctab %>% 
  mutate_at(vars(MRC_17_BUFF_48:MRC_WAT_205), funs(ifelse(. > 1 , 1, .))) %>% 
  mutate_at(vars(MRC_17_BUFF_48:MRC_WAT_205), funs(ifelse(. < 0, 0, .)))

rownames(ufctab) = rownames(ufcmat)

write.csv(ufctab, here("data/g98_UW_unifrac_dists_2021_check.csv"))

```

#### 4.5.ii Species level (averaged across individuals)

```{r}
# want to make a unifrac on the species level, so can average and re-run

# only mpala species
otuagg = mtusp[,c(2:167,169)] %>% group_by(Species) %>% mutate_at(vars(MOTU_1:MOTU_557), funs(mean))
otuagg=unique(otuagg)
otuotu_agg = otu_table(otuagg[,-167], taxa_are_rows = F)
comboseq_agg = phyloseq(treetree,otuotu_agg)

# create unifrac
set.seed(0)
ufc_agg = UniFrac(comboseq_agg, normalized=F, weighted =F) # is already normalized
hist(ufc_agg)

ufcmat_agg = as.matrix(ufc_agg)
rownames(ufcmat_agg)=otuagg$Species
colnames(ufcmat_agg)=otuagg$Species

write.csv(ufcmat_agg, here("data/g98_UW_unifrac_dists_sp_2021_check.csv"))
```


## 4. Phylogenetic Diversity!

Bring in tree information and use it to calculate phylogenetic diversity within the samples

```{r}
phylo_nems = treeNJ
phylo_nems=ladderize(phylo_nems)

# eliminate empty columns
RRA_present2 = RRA_present[, -c(which(colSums(RRA_present[,30:195])==0)+29)]
names(RRA_present2)

set.seed(0)
PDData_sum = data.frame(RRA_present2, ses.pd(dplyr::select(RRA_present2, MOTU_1:MOTU_491), tree=phylo_nems, include.root = F, runs=100, null.model="richness"))
# singletons get NAs

# Exploratory Plots:

# relationship between SR and PD
ggplot(PDData_sum, aes(x=ntaxa, y=pd.obs))+
  geom_point(aes(col=Species))+
  geom_smooth(aes(col=Species),method="lm", formula=y~log(x), se=F)+
  facet_wrap(~Species)

PDData_sum$Diet=factor(PDData_sum$Diet, levels=c("B","M","G"))

pdz = ggplot(PDData_sum[-which(PDData_sum$Species %in% c("Camel","Hippo")),], aes(x=Diet, y=pd.obs.z))+
  geom_hline(yintercept=-2, linetype="dotted")+
  geom_boxplot(data=PDData_sum[-which(PDData_sum$Species %in% c("Camel","Hippo")),], aes(x=Diet, y=pd.obs.z, col=Species), size=0.7, position=position_dodge(width=0.5), width=0.25, alpha=0.6)+
  geom_boxplot(fill=NA, col="gray50", size=1)+
  theme_bw(base_size=16)+
  labs(x="Diet", y="PD Z-Value", title="Standardized PD")+
  #scale_color_manual(values=c(animal_colors))+
  scale_color_manual(values=c(animal_colors[c(1,3:7)], animal_colors[c(8:11,13)],animal_colors[14:18]))+
  facet_wrap(~Feeding_Type, scales="free_x")+
  guides(color=F)


pd = ggplot(PDData_sum[-which(PDData_sum$Species %in% c("Camel","Hippo")),], aes(x=Diet, y=pd.obs))+
  geom_boxplot(data=PDData_sum[-which(PDData_sum$Species %in% c("Camel","Hippo")),], aes(x=Diet, y=pd.obs, col=Species), size=0.7, position=position_dodge(width=0.5), width=0.25, alpha=0.6)+
  geom_boxplot(fill=NA, col="gray50", size=1)+
  theme_bw(base_size=16)+
  labs(x="Diet", y="PD Z-Value", title="PD")+
  scale_color_manual(values=c(animal_colors[c(1,3:7)], animal_colors[c(8:11,13)],animal_colors[14:18]))+
  facet_wrap(~Feeding_Type, scales="free_x")+
  guides(color=F)

pdplot = gridExtra::grid.arrange(pd, pdz, ncol=2)
#ggsave(here("g98_plots/pdplot.png"),pdplot, device="png", dpi=300, width=10, height=5, units="in")

```

### Total PD

```{r}
PDData_3 = PDData_sum[-which(is.na(PDData_sum$pd.obs.p)),]
dim(PDData_3) # 213 individuals; 201 excluding CH

pdplot_sp = ggplot(PDData_3, aes(x=reorder(Species,-pd.obs,na.rm=T), y=pd.obs))+
  geom_jitter(width=0.01,height=0.01, alpha=0.5, aes(color=Species))+
  geom_boxplot(aes(fill=Species), color="gray50",outlier.shape=NA)+
  guides(fill=F,col=F)+
  theme_bw(base_size=16)+
  scale_color_manual(values=animal_colors)+
  scale_fill_manual(values=animal_colors)+
  theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0))+
  labs(x="Host Species",y="PD",title="Faith's PD")
pdplot_sp

# total PD by species
PD_spmod = lm(pd.obs~ Species, data=PDData_3)
summary(PD_spmod)
car::Anova(PD_spmod)
DHARMa::testResiduals(simulateResiduals(PD_spmod))
tab_model(PD_spmod)

empd = emmeans(PD_spmod, ~Species)
pairs(empd)%>%as.data.frame(.)%>%filter(p.value<0.05)

# PD diff by species
PD_spmodz = lm(pd.obs.z ~ Species, data=PDData_3)
summary(PD_spmodz)
DHARMa::testResiduals(simulateResiduals(PD_spmodz))
tab_model(PD_spmodz)
car::Anova(PD_spmodz)
# looks ok?

empdz= emmeans(PD_spmodz, ~Species)
pairs(empdz)%>%as.data.frame(.)%>%filter(p.value<0.05)

pdzplot_sp = ggplot(PDData_3, aes(x=reorder(Species,-pd.obs.z,na.rm=T), y=pd.obs.z))+
  geom_hline(yintercept = -2, linetype="dotted")+
  geom_jitter(width=0.01,height=0.01, alpha=0.5, aes(color=Species))+
  geom_boxplot(aes(fill=Species), color="gray50",outlier.shape=NA)+
  guides(fill=F,col=F)+
  theme_bw(base_size=16)+
  scale_color_manual(values=animal_colors)+
  scale_fill_manual(values=animal_colors)+
  theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0))+
  annotate(geom="text", x=5, y=-3.5, label="Lower than expected PD")+
  labs(x="Host Species",y="SES.PD",title="Standardized PD")

pdspplot = gridExtra::grid.arrange(pdplot_sp,pdzplot_sp,ncol=2)
#ggsave(here("g98_plots/pdsplots.png"),pdspplot, device="png", dpi=300, width=10, height=5, units="in")
```

### Investigate PD as a function of host characteristics

```{r}

inv.phylo = inverseA(tree, nodes="TIPS", scale=T)
prior_normal = list(G=list(G1=list(V=1,nu=0.02), G2=list(V=1,nu=0.02)),R=list(V=1,nu=0.02))

PDMCMC = as.data.frame(PDData_3[-which(PDData_3$Species %in% c("Camel","Hippo")),])
MPrichinfoMCMC$animal = MPrichinfoMCMC$Sample_ID
MPrichinfoMCMC=as.data.frame(MPrichinfoMCMC)
PDMCMC$phylo = PDMCMC$MSW93_Binomial
PDMCMC = as.data.frame(PDMCMC) 

# full model
set.seed(0)
model_simple2pd = MCMCglmm.updateable(pd.obs.z ~ Diet+ Feeding_Type + log(SocialGrpSize)+ log(HomeRange_km2)+ log(Body_Mass_kg),
                                      random=~phylo+MSW93_Binomial,
                                      family = "gaussian",
                                      ginverse=list(phylo=inv.phylo$Ainv),
                                      prior=prior_normal,
                                      data=PDMCMC,
                                      nitt=500000, burnin=10000, thin=250,
                                      verbose=F, pl=T)

sumpdz_full = make_MCMC_summary(model_simple2pd)
flextable:::knit_print.flextable(sumpdz_full)

```


```{r}

# faster model
set.seed(0)
model_simple2pd = MCMCglmm.updateable(pd.obs.z ~ Diet+ Feeding_Type + log(SocialGrpSize)+ log(HomeRange_km2)+ log(Body_Mass_kg),
                                      random=~phylo+MSW93_Binomial,
                                      family = "gaussian",
                                      ginverse=list(phylo=inv.phylo$Ainv),
                                      prior=prior_normal,
                                      data=PDMCMC,
                                      nitt=10000, burnin=5000, thin=50,
                                      verbose=F, pl=T)

dpd=dredge(model_simple2pd, rank="DIC")
dpd
# Feeding type best, but not much better than intercept alone

summary(model_simple2pd)

set.seed(0)
model_simple2pd_red = MCMCglmm.updateable(pd.obs.z ~ 1,
                                          random= ~ phylo+MSW93_Binomial,
                                          family = "gaussian",
                                          ginverse=list(phylo=inv.phylo$Ainv),
                                          prior=prior_normal,
                                          data=PDMCMC,
                                          nitt=500000, burnin=10000, thin=250,
                                          verbose=F, pl=T)


sumpdz = make_MCMC_summary(model_simple2pd_red)
flextable:::knit_print.flextable(sumpdz)
```


```{r}

#### Just PD
# slow model
set.seed(0)
model_simple2pd.obs = MCMCglmm.updateable(pd.obs ~Diet+Feeding_Type+log(SocialGrpSize)+log(HomeRange_km2)+ log(Body_Mass_kg), random=~phylo+MSW93_Binomial, family = "gaussian", ginverse=list(phylo=inv.phylo$Ainv), prior=prior_normal, data=PDMCMC, nitt=500000, burnin=10000, thin=250, verbose=F, pl=T)
summary(model_simple2pd.obs)

sumpdobs = make_MCMC_summary(model_simple2pd.obs)
flextable:::knit_print.flextable(sumpdobs)

```

```{r}
# fast version
set.seed(0)
model_simple2pd.obs = MCMCglmm.updateable(pd.obs ~Diet+Feeding_Type+log(SocialGrpSize)+log(HomeRange_km2)+ log(Body_Mass_kg), random=~phylo+MSW93_Binomial, family = "gaussian", ginverse=list(phylo=inv.phylo$Ainv), prior=prior_normal, data=PDMCMC, nitt=10000, burnin=5000, thin=50, verbose=F, pl=T)
summary(model_simple2pd.obs)

dpd2=dredge(model_simple2pd.obs, rank="DIC")
dpd2
# only intercept

set.seed(0)
model_simple2pd.obs.red = MCMCglmm.updateable(pd.obs ~1, random=~phylo+MSW93_Binomial, family = "gaussian", ginverse=list(phylo=inv.phylo$Ainv), prior=prior_normal, data=PDMCMC, nitt=500000, burnin=10000, thin=250, verbose=F, pl=T)

sumpdobs_red = make_MCMC_summary(model_simple2pd.obs.red)
flextable:::knit_print.flextable(sumpdobs)

```


## 5. Ordination (Parasite communities)

### 5.1 Create ordination objects using Jaccard distance

```{r}

# compare the MDS dimensions
ordMRC = metaMDS(dplyr::select(RRA_present2, MOTU_1:MOTU_491), k=3, binary=T, distance="jaccard",autotransform = F, trymax=100)
ordMRC2 = metaMDS(dplyr::select(RRA_present2, MOTU_1:MOTU_491), k=2, binary=T, distance="jaccard",autotransform = F, trymax=100)
ordMRC2 = metaMDS(dplyr::select(RRA_present2, MOTU_1:MOTU_491), k=2, binary=T, distance="jaccard",autotransform = F, trymax=100, previous.best = ordMRC2)

# determine k
stressplot(ordMRC)
stressplot(ordMRC2)

gof3 = goodness(ordMRC)
gof2 = goodness(ordMRC2)
plot(gof2 ~ gof3);abline(0,1) # very similar

plot(ordMRC);points(ordMRC, display="sites",cex=gof3*100)
plot(ordMRC2);points(ordMRC2, display="sites",cex=gof2*100) # probably two points benefit

pro3 = procrustes(ordMRC, ordMRC2)
plot(pro3) 
# three dimensions isn't that much better; will proceed with two


# 2D plot is probably more parsimonious
mdsplot = data.frame(x1 = ordMRC2$points[,1], x2=ordMRC2$points[,2], species = RRA_present$Species)

MRCord = ggplot(mdsplot, aes(x=x1, y=x2))+
  geom_point(aes(col=species), size=2)+
  geom_convexhull(aes(fill=species), alpha=0.2)+
  scale_fill_manual(values=animal_colors)+
  scale_color_manual(values=animal_colors)+
  theme_bw()+guides(fill=F, col=F)+
  annotate(geom="text", x=-6.5, y=-1.5, label=paste("Stress = ",round(ordMRC2$stress,2)))

MRCord

#ggsave("g98_plots/MRCord_k2_motu.png", MRCord,device="png")

```

### 5.2 Statistical tests

```{r}
envMRC=dplyr::select(RRA_present2, Species,Period)
# check NAs
envMRC$Period[which(is.na(envMRC$Period))]

dim(RRA_present2)

# Create distance matrix
MRCdist = vegdist(dplyr::select(RRA_present2, MOTU_1:MOTU_491), distance="jaccard", binary=T)
MRCdiv = dbrda(MRCdist ~ Species,  data=envMRC, permutations=999, method="jaccard")
MRCdiv0 = dbrda(MRCdist ~ 1,  data=envMRC, permutations=999, method="jaccard")

ordiR2step(MRCdiv0,MRCdiv)
anova(MRCdiv)

```

### Investigate species-level

```{r}

# To investigate other predictors, aggregate to the species level by averaging
MPagg = RRAagg[-which(RRAagg$Species%in%c("Hybrid zebra", "Camel","Hippo")),]

# ordinate the smaller DF
aggord = metaMDS(RRAagg[,-1], distance="jaccard", k=2, autotransform = F, binary=T)

# add data
MPenv_agg=left_join(MPagg, dplyr::select(RRA_present, Species, MSW93_Order:Feeding_Type))%>%unique()

# calculate distance matrix
MRCdist_agg = vegdist(MPagg[,-1], distance="jaccard", binary=T)

# Test for effects
MRCdiv_agg = dbrda(MRCdist_agg ~ Diet + Feeding_Type + log(Body_Mass_kg) + log(SocialGrpSize)+ log(HomeRange_km2), data=MPenv_agg, permutations=999, distance="jaccard")

MRCdiv_agg0 = dbrda(MRCdist_agg ~ 1, data=MPenv_agg, permutations=999, distance="jaccard")

anova(MRCdiv_agg, by="margin") # feeding type significant
ordistep(MRCdiv_agg, trace = F)
ordiR2step(MRCdiv_agg0, MRCdiv_agg, trace=F)
# Feeding Type is the only variable kept in

MRCdiv_agg_red =dbrda(MRCdist_agg ~ Feeding_Type, data=MPenv_agg, permutations=999, distance="jaccard")
anova(MRCdiv_agg_red)

# reduced model, but with consideration of taxonomy
MRCdbagg_red = dbrda(MRCdist_agg ~ Feeding_Type + MSW93_Order + MSW93_Family, data=MPenv_agg, permutations=999, distance="jaccard")
MRCdbagg_red
ordistep(MRCdbagg_red, direction = "both", trace=F, permutations = 1000, steps=1000)
# best model contains feeding type and order, rather than family

# create reduced final model
MRCdbagg_red = dbrda(MRCdist_agg ~ Feeding_Type + MSW93_Order, data=MPenv_agg, permutations=999, distance="jaccard")
anova(MRCdbagg_red, by="margin")
anova(MRCdbagg_red) # overall test


# plot results
agg_ord_plot = data.frame(Species=RRAagg$Species, x1=aggord$points[,1], x2=aggord$points[,2])
agg_ord_plot = left_join(agg_ord_plot, dplyr::select(RRA_present, Species, MSW93_Order:Feeding_Type))%>%unique()
head(agg_ord_plot)

#agg_ord_plot$Diet = factor(agg_ord_plot$Diet, levels=c("B","M","G"))

agg_ord_plot$Feeding_Type = factor(agg_ord_plot$Feeding_Type, levels=c("HGF","R","PR"))
library(stringr)
species_level_ord =ggplot(agg_ord_plot, aes(x=x1, y=x2))+
  geom_point(aes(col=MSW93_Order, shape=MSW93_Family), size=3, stroke=2)+
  geom_convexhull(aes(fill=Feeding_Type),color="gray",alpha=0.3)+guides(linetype=F)+
  scale_shape_manual(name="Family",values=c(1,3,5,7,12, 15, 16))+
  scale_color_manual(name="Order",values=c("black","gray70","gray90"))+
  scale_fill_manual(name="Gut Morphology",values=c("aquamarine4", "deepskyblue4","gray"), labels=c("Hindgut Fermenter","Ruminant", "Other"))+
  ggrepel::geom_text_repel(aes(label=str_sub(Species,1)), nudge_y=0, max.overlaps = 15)+
  theme_bw(base_size=14)+
  annotate(geom="text", x=2, y=-2, label=paste("Stress =", round(aggord$stress,2)))+
  labs(x="NMDS1",y="NMDS2")
species_level_ord

#ggsave("g98_plots/Species_Level_Ord_3.png", species_level_ord, device="png")


```

### 5.3 Tests using UniFrac Data

#### 5.3.i Full dataset

```{r}
# corresponding distance matrix
ufc_nems = read.csv(here("data/g98_UW_unifrac_dists_2021_check.csv"))
dim(ufc_nems)
#ufc_nems = ufcmat

# now use meta mds for visualization
unifmds = metaMDS(ufc_nems[,-1], autotransform = F,binary=T, dist="jaccard", k=2)

unifmdsdat = data.frame(x1=unifmds$points[,1], x2=unifmds$points[,2], sample=ufc_nems$X)
unifmdsdat = left_join(unifmdsdat, RRA_present, by=c("sample"="Sample_ID"))

levels(as.factor(unifmdsdat$Species))

# plot
mds_uf_sp =ggplot(unifmdsdat, aes(x=x1, y=x2))+
  geom_point(aes(col=Species))+
  geom_convexhull(aes(fill=Species), alpha=0.2)+
  scale_fill_manual(values=animal_colors)+
  scale_color_manual(values=animal_colors)+
  theme_bw(base_size=18)+guides(fill=F, col=F)+
  annotate(geom="text", x=-0.2, y=-0.6, label=paste("Stress =", round(unifmds$stress,2)), size=5)+
  labs(x="NMDS 1", y="NMDS 2")# stress is 0.105

mds_uf_sp

ggplot(unifmdsdat, aes(x=x1, y=x2))+
  geom_jitter(aes(col=Species))+
  geom_convexhull(aes(fill=Species), alpha=0.2)+
  scale_fill_manual(values=animal_colors)+
  scale_color_manual(values=animal_colors)+
  theme_bw(base_size=18)+guides(fill=F, col=F)+
  labs(x="NMDS 1", y="NMDS 2")+facet_wrap(~Species)

#ggsave("g98_plots/MDS_with_unifrac_2021.png", mds_uf_sp, device="png")


# add host info
unifmdsdat = left_join(unifmdsdat, dplyr::select(RRA_present, Species, MSW93_Order:Feeding_Type))%>%unique()

unif_feeding = ggplot(unifmdsdat[-which(unifmdsdat$Species %in% c("Hybrid zebra", "Camel","Hippo")),], aes(x=x1, y=x2))+
  geom_point(aes(col=Feeding_Type))+
  geom_convexhull(aes(col=Feeding_Type, fill=Feeding_Type, group=MSW93_Family), alpha=0.1)+
  theme_bw()
unif_feeding

#ggsave("g98_plots/Feeding_MDS_unif.png", unif_feeding, device="png")

# conduct test

# Create distance matrix
MRCdiv_uf = dbrda(ufc_nems[,-1] ~ Species,  data=unifmdsdat, permutations=999)
MRCdiv0_uf = dbrda(ufc_nems[,-1] ~ 1,  data=unifmdsdat, permutations=999)

ordiR2step(MRCdiv0_uf,MRCdiv_uf)
MRCdiv_uf
anova(MRCdiv_uf)

```

#### 5.3.ii Species level data

```{r}
# species level
# exclude camel and hippo
ufcmat_agg2 = ufcmat_agg[ -which( rownames(ufcmat_agg) %in% c("Camel","Hippo")),
                          -which(rownames(ufcmat_agg) %in% c("Camel","Hippo")) ]

# run MDS
ufcaggmds = metaMDS(ufcmat_agg, k=2, binary=T, dist="jaccard", autotransform = F)

# create dataframe
unifmdsdat2 = data.frame(x1 = ufcaggmds$points[,1],
                         x2 = ufcaggmds$points[,2],
                         Species = rownames(ufcmat_agg))

# add species information
unifmdsdat2 = left_join(unifmdsdat2,
                        dplyr::select(RRA_present, Species, MSW93_Order:Feeding_Type)) %>%
  unique()

# run MDS on full df for plot
ufcaggmds = metaMDS( ufcmat_agg,
                     k=2,
                     binary=T,
                     dist="jaccard",
                     autotransform = F)

# create data frame
unifmdsdat3 = data.frame(x1=ufcaggmds$points[,1],
                         x2=ufcaggmds$points[,2],
                         Species=rownames(ufcmat_agg))

# add species information
unifmdsdat3 = left_join(unifmdsdat3,
                        dplyr::select(RRA_present, Species, MSW93_Order:Feeding_Type)) %>%
  unique()

# plot
mds_uf_sp2 =ggplot(unifmdsdat2, aes(x=x1, y=x2))+
  geom_point(aes(col=Feeding_Type, shape=MSW93_Family),
             size=3,
             stroke=2)+
  geom_convexhull(aes(fill=Feeding_Type),
                  alpha=0.2)+
  ggrepel::geom_text_repel(aes(label=Species),
                           size=5)+
  scale_shape_manual(values=c(1,3,5,7,12, 15, 16))+
  guides(shape=F)+
  theme_bw(base_size=14)+
  labs(x="NMDS1", y="NMDS2")

mds_uf_sp2

#ggsave("g98_plots/unifrac_MDS_species.png", mds_uf_sp2, device="png")

# test using all covariates
MRCdiv_agg_ufc = dbrda(ufcmat_agg2 ~ Diet + Feeding_Type + log(Body_Mass_kg) + log(SocialGrpSize)+ log(HomeRange_km2),
                       data=unifmdsdat2[ -which(unifmdsdat2$Species %in% c("Camel","Hippo")), ],
                       permutations=999,
                       method="jaccard")

anova(MRCdiv_agg_ufc, by="margin") # feeding type
ordistep(MRCdiv_agg_ufc, trace=F) # feeding type

# best predictor
MRCdiv_agg_ufc = dbrda(ufcmat_agg2 ~ Feeding_Type,
                       data=unifmdsdat2[ -which(unifmdsdat2$Species %in% c("Camel","Hippo")), ],
                       permutations=999,
                       method="jaccard")

# null model
MRCdiv_agg_ufc0 = dbrda(ufcmat_agg2 ~ 1,
                        data=unifmdsdat2[ -which(unifmdsdat2$Species %in% c("Camel","Hippo")), ],
                        permutations=999,
                        method="jaccard")

# compare R2
ordiR2step(MRCdiv_agg_ufc0, MRCdiv_agg_ufc) #37

# add taxonomic data
MRCdiv_agg_ufc = dbrda(ufcmat_agg2 ~ Feeding_Type + MSW93_Family + MSW93_Order,
                       data = unifmdsdat2[-which(unifmdsdat2$Species %in% c("Camel","Hippo")),],
                       permutations=999,
                       method="jaccard")

ordistep(MRCdiv_agg_ufc, trace=F) # Feeding type and order --> 75%

# best model
MRCdiv_agg_ufc = dbrda(ufcmat_agg2 ~ Feeding_Type + MSW93_Order,
                       data = unifmdsdat2[-which(unifmdsdat2$Species %in% c("Camel","Hippo")), ],
                       permutations=999,
                       method="jaccard")

anova(MRCdiv_agg_ufc, by="margin")

# plot
# species_level_ord_uf =ggplot(unifmdsdat2, aes(x=x1, y=x2))+
#   geom_point(aes(col=MSW93_Order, shape=MSW93_Family), size=3, stroke=2)+
#   geom_convexhull(aes(fill=Feeding_Type),color="gray",alpha=0.3)+guides(linetype=F)+
#   scale_shape_manual(name="Family",values=c(1,3,5,7,12, 15, 16))+
#   scale_color_manual(name="Order",values=c("black","gray40","gray70"))+
#   scale_fill_manual(name="Gut Morphology",values=c("aquamarine4", "deepskyblue4"))+
#   ggrepel::geom_text_repel(aes(label=str_sub(Species,1)), nudge_y=0)+
#   theme_bw(base_size=14)+
#   #guides(shape="Family",color="Order", fill="Gut Morphology")+
#   labs(x="NMDS1",y="NMDS2")
# species_level_ord_uf

# with cam and hip
unifmdsdat3$Feeding_Type = factor(unifmdsdat3$Feeding_Type, levels=c("HGF","R","PR"))

species_level_ord_uf =ggplot(unifmdsdat3, aes(x=x1, y=x2))+
  geom_point(aes(col=MSW93_Order, shape=MSW93_Family),
             size=3,
             stroke=2)+
  geom_convexhull(aes(fill=Feeding_Type),
                  color="gray",alpha=0.3)+
  guides(linetype=F)+
  scale_shape_manual(name="Family",
                     values=c(1,3,5,7,12, 15, 16))+
  scale_color_manual(name="Order",
                     values=c("black","gray70","gray90"))+
  scale_fill_manual(name="Gut Morphology",
                    values=c("aquamarine4", "deepskyblue4","gray"),
                    labels=c("Hindgut Fermenter", "Ruminant","Other"))+
  ggrepel::geom_text_repel(aes(label=str_sub(Species,1)),
                           nudge_y=0,
                           max.overlaps = 13)+
  theme_bw(base_size=14)+
  annotate(geom="text", x=-0.2, y=-0.32,
           label=paste("Stress =",round(ufcaggmds$stress,2)))+
  #guides(shape="Family",color="Order", fill="Gut Morphology")+
  labs(x="NMDS1",y="NMDS2")

species_level_ord_uf

#ggsave(here("g98_plots/Species_Level_Ord_uf_2021.png"), species_level_ord_uf, device="png")
#ggsave(here("g98_plots/Species_Level_Ord_uf_2021.pdf"), species_level_ord_uf, device="pdf")


```


### 6.2 Include taxonomic information on MOTU mds (not available for Unifrac)

```{r}
# Extract specific MOTUS and determine weights on MDS
motu_wts = as.data.frame(ordMRC2$species)
motu_wts$MOTU =row.names(motu_wts)

motu_wts = unique(left_join(motu_wts, MOTU2))
motu_wts$name= motu_wts$name2
motu_wts$name = as.factor(motu_wts$name)


# task -- find centroid for each parasite family.
psitecent = motu_wts %>% group_by(name)%>% summarize_at(vars(MDS1:MDS2), funs(mean(.,na.rm=T)))
#plotMDSMRC2 = left_join(plotMDSMRC, colorsp, by=c("Species"="Host"))


colors_for_plot= c("dodgerblue", "magenta3","green3","maroon1", "maroon3", "maroon", "dodgerblue3",  "deepskyblue3", "dodgerblue4","gold3", "cadetblue1", "greenyellow", "green4", "deepskyblue","lightsteelblue3",  "gold", "darkorchid4", "lightskyblue2", "darkorchid1")

colors_for_plot2 = c("dodgerblue", "magenta3","green3","maroon1", "maroon3", "dodgerblue3", "deepskyblue3", "gold3",  "dodgerblue4", "cadetblue1", "maroon", "greenyellow", "green4", "deepskyblue",  "darkorchid4", "gold", "lightsteelblue3",  "darkorchid1",  "lightskyblue2")

mdsplot_motu = psitecent %>%
  ggplot(aes(x=-MDS1, y=MDS2))+
  theme_gray(base_size=12)+
  geom_jitter(data=mdsplot, aes(x=-x1,y=x2,col=species), size=2, stroke=1.5,width=0.1, height=0.1, alpha=0.5)+
  geom_convexhull(data=mdsplot, aes(x=-x1,y=x2, fill=species), alpha=0.3)+
  scale_fill_manual(values=animal_colors)+
  scale_color_manual(values=animal_colors)+
  geom_segment(aes(xend=0, yend=0))+
  ggrepel::geom_text_repel(aes(label=name), max.overlaps = 15)+
  labs(y="NMDS 2", x="NMDS 1")+
  annotate(geom="text", x = -6.5, y=-1.85, label=paste("Stress =", round(ordMRC2$stress,2)), size=5)+
  theme_bw(base_size=18)+guides(fill=F, col=F)



psitecent %>%
  ggplot(aes(x=-MDS1, y=MDS2))+
  theme_gray(base_size=12)+
  geom_jitter(data=mdsplot, aes(x=-x1,y=x2,col=species), size=2, stroke=1.5,width=0.1, height=0.1, alpha=0.5)+
  geom_convexhull(data=mdsplot, aes(x=-x1,y=x2, fill=species), alpha=0.3)+
  scale_fill_manual(values=animal_colors)+
  scale_color_manual(values=animal_colors)+
  geom_segment(aes(xend=0, yend=0))+
  facet_wrap(~mdsplot$species)+
  theme_bw(base_size=14)+
  guides(fill=F, col=F)+
  facet_wrap(~species) #colors aren't correct ughh

mdsplot_motu

#ggsave("g98_plots/MDS_MOTU_withPsites_2021.png", mdsplot_motu, device="png")
# for editing in illustrator:
#ggsave(here("g98_plots/MDS_MOTU_withPsites_fam_2021.pdf"), mdsplot_motu, device="pdf")



```




## 7. Creating a network

### 7.1 Setup

```{r}
# Prevalence of each motu
# Mutate RRA values < 1% 
RRA3 = RRA2 %>% mutate_at(vars(MOTU_1:MOTU_557), funs(ifelse(.<0.01,0,.)))

# Calculate average RRA for each species and mOTU combination
avgRRA = RRA2 %>%
  group_by(Species) %>%
  summarize_at(vars(MOTU_1:MOTU_557), funs(mean)) %>% 
  ungroup()

nRRA = RRA2 %>%
  group_by(Species) %>%
  summarize(n=n()) %>% 
  ungroup %>% 
  left_join(avgRRA)
nRRA$r = specnumber(nRRA[,3:168])
ggplot(nRRA, aes(x=n, y=r))+
  geom_point()
cor.test(nRRA$r, nRRA$n)
# There is a slight correlation between species richness and n


samp_mat_1 = avgRRA[, -c(which(colSums(avgRRA[,-1]) == 0) + 1)]

levelplot(as.matrix(samp_mat_1[,-1]),  col.regions = gray(0:100/100))

unlist(samp_mat_1[,-1])[unlist(samp_mat_1[,-1])>0.001]%>%hist() # avg RRA > 0.01 when present

```

### 7.2 Plotting
```{r}

# need to put all in one column
mat_1_long <- gather(samp_mat_1, key=OTU, value=RRA, -Species)
head(mat_1_long)

# remove very rare or zero
#mat_1_long <- mat_1_long[-which(mat_1_long$RRA == 0),]

g <- graph.data.frame(mat_1_long, directed=FALSE)

V(g)$type <- bipartite_mapping(g)$type 
V(g)$color <- ifelse(V(g)$type, "lightblue", "salmon")
V(g)$shape <- ifelse(V(g)$type, "square", "circle")
E(g)$color <- "lightgray"
V(g)$frame.color <-  "white"
V(g)$label.dist <-0.5
E(g)$weight = E(g)$RRA


plot(g, vertex.label.cex = 0.8, vertex.size=ifelse(V(g)$type,1,7), vertex.label.color = ifelse(V(g)$type,"gray","black"), edge.width=E(g)$weight)


# use at least 5 samples
RRA_present %>% group_by(Species) %>% summarize(n()) # no kudu or waterbuck
#mat_2_long = mat_1_long[-which(mat_1_long$Species %in% c("Waterbuck","Kudu")),]
mat_2_long = mat_1_long

g <- graph.data.frame(mat_2_long, directed=FALSE)

V(g)$type <- bipartite_mapping(g)$type 
V(g)$color <- ifelse(V(g)$type, "lightblue", "salmon")
V(g)$shape <- ifelse(V(g)$type, "square", "circle")
E(g)$color <- "lightgray"
V(g)$frame.color <-  "white"
V(g)$label.dist <-0.5
E(g)$weight = E(g)$RRA


plot(g, vertex.label.cex = 0.8, vertex.size=ifelse(V(g)$type,1,7), vertex.label.color = ifelse(V(g)$type,"gray","black"), edge.width=E(g)$weight)

```

### 7.3 Labeling MOTUs by parasite group


```{r}

motu98 = data.frame(OTU = paste("MOTU_",motuinfo$cluster_sumatra98,sep=""),
                    rank = motuinfo$rank_ok,
                    sp = motuinfo$scientific_name_ok,
                    sp2 = motuinfo$unknown)

motu98 %>% group_by(OTU) %>% summarize(n()) %>% ungroup()

test = left_join(mat_2_long, motu98)

# takes a second
test=unique(test)

# which MOTUs have different sp names
getuni = test %>% group_by(OTU,rank,sp,sp2) %>% summarize(n())

# manually get the taxon
# first create a new column
test$sp_f = rep("",length(test$sp))

test=as.data.frame(test)


test[which(test$OTU=="MOTU_1"),c(5,6,7)]$sp_f = "Cooperia"
test[which(test$OTU=="MOTU_2"),c(2,4,5,6,7)]$sp_f = "Murshidia_linstowi"
if(length(which(test$OTU=="MOTU_3"))!=0){
  test[which(test$OTU=="MOTU_3"),c(5,6,7)]$sp_f = "Oesophagostomum_dentatum"}
test[which(test$OTU=="MOTU_4"),c(5,6,7)]$sp_f = "Haemonchus"
test[which(test$OTU=="MOTU_5"),c(5,6,7)]$sp_f = "Cooperia_fuelleborni"
test[which(test$OTU=="MOTU_6"),c(5,6,7)]$sp_f = "Coronocyclus_labratus"
test[which(test$OTU=="MOTU_7"),c(5,6,7)]$sp_f = "Strongyloidea"
test[which(test$OTU=="MOTU_8"),c(5,6,7)]$sp_f = "Murshidia_africana"
test[which(test$OTU=="MOTU_9"),c(5,6,7)]$sp_f = "Oesophagostomum_dentatum"
test[which(test$OTU=="MOTU_10"),c(5,6,7)]$sp_f = "Cooperia"
test[which(test$OTU=="MOTU_11"),c(2,4,5,6,7)]$sp_f = "Murshidia_africana"
test[which(test$OTU=="MOTU_12"),c(2,4,5,6,7)]$sp_f = "Strongyloidea"
test[which(test$OTU=="MOTU_13"),c(2,4,5,6,7)]$sp_f = "Trichostrongylus vitrinus"
test[which(test$OTU=="MOTU_14"),c(2,4,5,6,7)]$sp_f = "Trichostrongylus"
test[which(test$OTU=="MOTU_15"),c(2,4,5,6,7)]$sp_f = "Cooperia pectinata"
test[which(test$OTU=="MOTU_16"),c(2,4,5,6,7)]$sp_f = "Trichostrongylus"
test[which(test$OTU=="MOTU_17"),c(2,4,5,6,7)]$sp_f = "Trichostrongyloidea"
test[which(test$OTU=="MOTU_18"),c(2,4,5,6,7)]$sp_f = "Murshidia_linstowi"
test[which(test$OTU=="MOTU_19"),c(2,4,5,6,7)]$sp_f = "Parapoteriostomum_euproctus"
test[which(test$OTU=="MOTU_20"),c(2,4,5,6,7)]$sp_f = "Cylicostephanus_minutus"
test[which(test$OTU=="MOTU_21"),c(2,4,5,6,7)]$sp_f = "Murshidia_longicaudata"
test[which(test$OTU=="MOTU_22"),c(2,4,5,6,7)]$sp_f = "Cylicostephanus_bidentatus"
test[which(test$OTU=="MOTU_23"),c(2,4,5,6,7)]$sp_f = "Cloacinidae"
test[which(test$OTU=="MOTU_24"),c(2,4,5,6,7)]$sp_f = "Cylicocyclus_adersi"
test[which(test$OTU=="MOTU_25"),c(2,4,5,6,7)]$sp_f = "Haemonchus"
test[which(test$OTU=="MOTU_26"),c(2,4,5,6,7)]$sp_f = "Strongyloidea"
test[which(test$OTU=="MOTU_27"),c(2,4,5,6,7)]$sp_f = "Khalilia_sameera"
test[which(test$OTU=="MOTU_28"),c(2,4,5,6,7)]$sp_f = "Quilonia_africana"
test[which(test$OTU=="MOTU_29"),c(2,4,5,6,7)]$sp_f = "Haemonchus"
test[which(test$OTU=="MOTU_30"),c(2,4,5,6,7)]$sp_f = "Oesophagostomum_radiatum"
test[which(test$OTU=="MOTU_31"),c(2,4,5,6,7)]$sp_f = "Strongylus_edentatus"
test[which(test$OTU=="MOTU_32"),c(2,4,5,6,7)]$sp_f = "Ostertagia_nianqingtanggulansis"
test[which(test$OTU=="MOTU_33"),c(2,4,5,6,7)]$sp_f = "Trichostrongyloidea"
test[which(test$OTU=="MOTU_34"),c(2,4,5,6,7)]$sp_f = "Skrjabinodentus_caragandicus"
test[which(test$OTU=="MOTU_35"),c(2,4,5,6,7)]$sp_f = "Cooperia"
test[which(test$OTU=="MOTU_36"),c(2,4,5,6,7)]$sp_f = "Strongylus_equinus"
test[which(test$OTU=="MOTU_37"),c(2,4,5,6,7)]$sp_f = "Trichostrongylus_axei"
test[which(test$OTU=="MOTU_38"),c(2,4,5,6,7)]$sp_f = "Trichostrongylus_colubriformis"
test[which(test$OTU=="MOTU_39"),c(2,4,5,6,7)]$sp_f = "Craterostomum_acuticaudatum"
test[which(test$OTU=="MOTU_40"),c(2,4,5,6,7)]$sp_f = "Trichostrongylidae"
test[which(test$OTU=="MOTU_41"),c(2,4,5,6,7)]$sp_f = "Quilonia_africana"
test[which(test$OTU=="MOTU_42"),c(2,4,5,6,7)]$sp_f = "Trichostrongylus"
test[which(test$OTU=="MOTU_43"),c(2,4,5,6,7)]$sp_f = "Coronocyclus_labiatus"
test[which(test$OTU=="MOTU_44"),c(2,4,5,6,7)]$sp_f = "Strongyloidea"
test[which(test$OTU=="MOTU_45"),c(2,4,5,6,7)]$sp_f = "Skrabinodentus_caragandicus"
test[which(test$OTU=="MOTU_46"),c(2,4,5,6,7)]$sp_f = "Coronocyclus_labiatus"
test[which(test$OTU=="MOTU_47"),c(2,4,5,6,7)]$sp_f = "Trichostrongylinae"
test[which(test$OTU=="MOTU_48"),c(2,4,5,6,7)]$sp_f = "Cooperia_spatulata"
test[which(test$OTU=="MOTU_49"),c(2,4,5,6,7)]$sp_f = "Strongylus_vulgaris"
test[which(test$OTU=="MOTU_50"),c(2,4,5,6,7)]$sp_f = "Triodontophorus_nipponicus"
test[which(test$OTU=="MOTU_51"),c(2,4,5,6,7)]$sp_f = "Quilonia_africana"
test[which(test$OTU=="MOTU_52"),c(2,4,5,6,7)]$sp_f = "Haemonchus_contortus"
#test[which(test$OTU=="MOTU_53"),c(2,4,5,6,7)]$sp_f = "Strongyloidea"
test[which(test$OTU=="MOTU_54"),c(2,4,5,6,7)]$sp_f = "Trichostrongylus_colubriformis"
test[which(test$OTU=="MOTU_55"),c(2,4,5,6,7)]$sp_f = "Cloacinidae"
test[which(test$OTU=="MOTU_56"),c(2,4,5,6,7)]$sp_f = "Trichostrongylus_rugatus"
test[which(test$OTU=="MOTU_57"),c(2,4,5,6,7)]$sp_f = "Coronocyclus_labratus"
if(length(which(test$OTU=="MOTU_58"))!=0){
test[which(test$OTU=="MOTU_58"),c(2,4,5,6,7)]$sp_f = "Cooperia_fuelleborni"}
test[which(test$OTU=="MOTU_59"),c(2,4,5,6,7)]$sp_f = "Strongyloidea"
test[which(test$OTU=="MOTU_60"),c(2,4,5,6,7)]$sp_f = "Murshidia_africana"
test[which(test$OTU=="MOTU_61"),c(2,4,5,6,7)]$sp_f = "Trichostrongylus"
test[which(test$OTU=="MOTU_62"),c(2,4,5,6,7)]$sp_f = "Cyclicocyclus_auriculatus"
test[which(test$OTU=="MOTU_63"),c(2,4,5,6,7)]$sp_f = "Cyclicocyclus_auriculatus"
test[which(test$OTU=="MOTU_64"),c(2,4,5,6,7)]$sp_f = "Trichostrongylidae"
if(length(which(test$OTU=="MOTU_66"))!=0){
test[which(test$OTU=="MOTU_66"),c(2,4,5,6,7)]$sp_f = "Coronocyclus_labiatus"}
test[which(test$OTU=="MOTU_67"),c(2,4,5,6,7)]$sp_f = "Haemonchus_contortus"
test[which(test$OTU=="MOTU_68"),c(2,4,5,6,7)]$sp_f = "Oesophagostomum"
if(length(which(test$OTU=="MOTU_69"))!=0){
test[which(test$OTU=="MOTU_69"),c(2,4,5,6,7)]$sp_f = "Trichostrongylus"}
test[which(test$OTU=="MOTU_70"),c(2,4,5,6,7)]$sp_f = "Skrabinodentus_caragandicus"
test[which(test$OTU=="MOTU_71"),c(2,4,5,6,7)]$sp_f = "Strongyloidea"
test[which(test$OTU=="MOTU_72"),c(2,4,5,6,7)]$sp_f = "Strongylus_vulgaris"
if(length(which(test$OTU=="MOTU_3"))!=0){
test[which(test$OTU=="MOTU_74"),c(2,4,5,6,7)]$sp_f = "Trichostrongylus_vitrinus"}
test[which(test$OTU=="MOTU_75"),c(2,4,5,6,7)]$sp_f = "Trichostrongylidae"
test[which(test$OTU=="MOTU_76"),c(2,4,5,6,7)]$sp_f = "Strongyloidea"
test[which(test$OTU=="MOTU_77"),c(2,4,5,6,7)]$sp_f = "Strongylus_equinus"
test[which(test$OTU=="MOTU_78"),c(2,4,5,6,7)]$sp_f = "Trichostrongylus"
test[which(test$OTU=="MOTU_79"),c(2,4,5,6,7)]$sp_f = "Trichostrongylus_rugatus"
test[which(test$OTU=="MOTU_81"),c(2,4,5,6,7)]$sp_f = "Cooperia"
test[which(test$OTU=="MOTU_82"),c(2,4,5,6,7)]$sp_f = "Coronocyclus_labiatus"
test[which(test$OTU=="MOTU_83"),c(2,4,5,6,7)]$sp_f = "Cooperia_fuelleborni"
test[which(test$OTU=="MOTU_84"),c(2,4,5,6,7)]$sp_f = "Strongyloidea"
test[which(test$OTU=="MOTU_85"),c(2,4,5,6,7)]$sp_f = "Trichostrongylidae"
test[which(test$OTU=="MOTU_87"),c(2,4,5,6,7)]$sp_f = "Cyclicostephanus_minutus"
test[which(test$OTU=="MOTU_89"),c(2,4,5,6,7)]$sp_f = "Strongyloidea"
test[which(test$OTU=="MOTU_90"),c(2,4,5,6,7)]$sp_f = "Strongyloidea"
test[which(test$OTU=="MOTU_91"),c(2,4,5,6,7)]$sp_f = "Triodontophorus_serratus"
test[which(test$OTU=="MOTU_92"),c(2,4,5,6,7)]$sp_f = "Strongylus_edentatus"
if(length(which(test$OTU=="MOTU_93"))!=0){
test[which(test$OTU=="MOTU_93"),c(2,4,5,6,7)]$sp_f = "Trichostrongylus"}
test[which(test$OTU=="MOTU_94"),c(2,4,5,6,7)]$sp_f = "Poteriostomum_imparidentatum"
if(length(which(test$OTU=="MOTU_97"))!=0){
test[which(test$OTU=="MOTU_97"),c(2,4,5,6,7)]$sp_f = "Murshidia_linstowi"}
test[which(test$OTU=="MOTU_100"),c(2,4,5,6,7)]$sp_f = "Trichostrongylus_rugatus"
test[which(test$OTU=="MOTU_101"),c(2,4,5,6,7)]$sp_f = "Murshidia_linstowi"
test[which(test$OTU=="MOTU_102"),c(2,4,5,6,7)]$sp_f = "Strongyloidea"
if(length(which(test$OTU=="MOTU_108"))!=0){
test[which(test$OTU=="MOTU_108"),c(2,4,5,6,7)]$sp_f = "Murshidia"}
test[which(test$OTU=="MOTU_109"),c(2,4,5,6,7)]$sp_f = "Trichostrongylidae"
test[which(test$OTU=="MOTU_110"),c(2,4,5,6,7)]$sp_f = "Coronocyclus_labratus"
if(length(which(test$OTU=="MOTU_111"))!=0){
test[which(test$OTU=="MOTU_111"),c(2,4,5,6,7)]$sp_f = "Strongyloidea"}
test[which(test$OTU=="MOTU_112"),c(2,4,5,6,7)]$sp_f = "Trichostrongylidae"
test[which(test$OTU=="MOTU_114"),c(2,4,5,6,7)]$sp_f = "Coronocyclus_labiatus"
if(length(which(test$OTU=="MOTU_121"))!=0){
test[which(test$OTU=="MOTU_121"),c(2,4,5,6,7)]$sp_f = "Strongyloidea"}
test[which(test$OTU=="MOTU_123"),c(2,4,5,6,7)]$sp_f = "Trichostrongylus_colubriformis"
test[which(test$OTU=="MOTU_124"),c(2,4,5,6,7)]$sp_f = "Ostertagia_nianqingtanggulansis"
test[which(test$OTU=="MOTU_126"),c(2,4,5,6,7)]$sp_f = "Trichostrongylus"
test[which(test$OTU=="MOTU_129"),c(2,4,5,6,7)]$sp_f = "Bunostomum_phlebotomum"
if(length(which(test$OTU=="MOTU_143"))!=0){
test[which(test$OTU=="MOTU_143"),c(2,4,5,6,7)]$sp_f = "Cooperia_fuelleborni"}
test[which(test$OTU=="MOTU_146"),c(2,4,5,6,7)]$sp_f = "Trichostrongylus"
if(length(which(test$OTU=="MOTU_148"))!=0){
test[which(test$OTU=="MOTU_148"),c(2,4,5,6,7)]$sp_f = "Oesophagostomum_dentatum"}
test[which(test$OTU=="MOTU_149"),c(2,4,5,6,7)]$sp_f = "Coronocyclus_labiatus"
if(length(which(test$OTU=="MOTU_153"))!=0){
test[which(test$OTU=="MOTU_153"),c(2,4,5,6,7)]$sp_f = "Umingmakstrongylus_pallikuukensis"}
test[which(test$OTU=="MOTU_157"),c(2,4,5,6,7)]$sp_f = "Coronocyclus_labiatus"
test[which(test$OTU=="MOTU_162"),c(2,4,5,6,7)]$sp_f = "Haemonchus_contortus"
test[which(test$OTU=="MOTU_163"),c(2,4,5,6,7)]$sp_f = "Cooperia_fuelleborni"
test[which(test$OTU=="MOTU_165"),c(2,4,5,6,7)]$sp_f = "Bunostomum"
test[which(test$OTU=="MOTU_172"),c(2,4,5,6,7)]$sp_f = "Coronocyclus_labiatus"
test[which(test$OTU=="MOTU_173"),c(2,4,5,6,7)]$sp_f = "Coronocyclus_labiatus"
test[which(test$OTU=="MOTU_192"),c(2,4,5,6,7)]$sp_f = "Cooperia_fuelleborni"
test[which(test$OTU=="MOTU_194"),c(2,4,5,6,7)]$sp_f = "Coronocyclus_labiatus"
if(length(which(test$OTU=="MOTU_205"))!=0){
test[which(test$OTU=="MOTU_205"),c(2,4,5,6,7)]$sp_f = "Necator"}
test[which(test$OTU=="MOTU_211"),c(2,4,5,6,7)]$sp_f = "Trichostrongylus"
test[which(test$OTU=="MOTU_216"),c(2,4,5,6,7)]$sp_f = "Cooperia_pectinata"
test[which(test$OTU=="MOTU_229"),c(2,4,5,6,7)]$sp_f = "Haemonchus_bedfordi"
test[which(test$OTU=="MOTU_288"),c(2,4,5,6,7)]$sp_f = "Cylicocyclus_elongatus"
test[which(test$OTU=="MOTU_292"),c(2,4,5,6,7)]$sp_f = "Cooperia_curticei"
test[which(test$OTU=="MOTU_293"),c(2,4,5,6,7)]$sp_f = "Trichostrongylus_vitrinus"
if(length(which(test$OTU=="MOTU_423"))!=0){
test[which(test$OTU=="MOTU_423"),c(2,4,5,6,7)]$sp_f = "Cooperia_curticei"}
test[which(test$OTU=="MOTU_337"),c(2,4,5,6,7)]$sp_f = "Quilonia_africana"
if(length(which(test$OTU=="MOTU_486"))!=0){
test[which(test$OTU=="MOTU_486"),c(2,4,5,6,7)]$sp_f = "Cooperia"}
test[which(test$OTU=="MOTU_491"),c(2,4,5,6,7)]$sp_f = "Coronocyclus_coronatus"
if(length(which(test$OTU=="MOTU_557"))!=0){
test[which(test$OTU=="MOTU_557"),c(2,4,5,6,7)]$sp_f = "Oesophagostomum_dentatum"}


test %>%
  group_by(OTU,sp_f) %>%
  summarize(n()) %>% 
  ungroup()

test2 = (unique(test[,-4]))

test2 %>%
  group_by(OTU,sp_f) %>% summarize(n())
dim(test2) 


# need to make a column with sp and motu
test2$sp_f2 = paste(test2$sp_f,str_remove(test2$OTU, "MOTU_"),sep="_")

# unique values
test2 = test2 %>%
  dplyr::select(Species, sp_f2, OTU, RRA) %>%
  unique()


test3 = test2[-which(test2$RRA<0.001),] # removed extra cow -- elephant
#test3=test2


new_mat = test3[,c(1,2,4)]
g <- graph.data.frame(new_mat, directed=FALSE)

V(g)$type <- bipartite_mapping(g)$type 
V(g)$color <- ifelse(V(g)$type, "lightblue", "salmon")
V(g)$shape <- ifelse(V(g)$type, "square", "circle")
E(g)$color <- "lightgray"
V(g)$frame.color <-  "white"
V(g)$label.dist <-0.5


plot(g, vertex.label.cex = 1.2, vertex.size=ifelse(V(g)$type,3,8), vertex.label.color = ifelse(V(g)$type,"gray","black"))


####

# haven't written updated df yet
#write.csv(test2, "data_for_network_graph.csv", row.names=F)
```



```{r}
#test2 = read.csv("data_for_network_graph.csv")
head(test3)

# rearrange
test3 = test3[,c(1,2,4,3)]

g <- graph.data.frame(test3, directed=FALSE)

V(g)$type <- bipartite_mapping(g)$type 
V(g)$color <- ifelse(V(g)$type, "lightblue", "salmon")
V(g)$shape <- ifelse(V(g)$type, "square", "circle")
E(g)$color <- "lightgray"
V(g)$frame.color <-  "white"
V(g)$label.dist <-0.5
E(g)$weight = E(g)$RRA

plot(g, vertex.label.cex = 0.9, vertex.size=ifelse(V(g)$type,3,8), vertex.label.color = ifelse(V(g)$type,"gray","black"), edge.width =E(g)$weight, layout=layout_nicely)

```


### 7.3 Create clusters based on Jaccard distance

```{r}

test4 = test3 %>% dplyr::select(Species, OTU, RRA) %>% spread(key=OTU, value=RRA, fill=0)

allRRA = unlist(test4[,-1])
hist(log(allRRA[allRRA>0]))

test5 = as.matrix(test4[,-1])
row.names(test5)=test4$Species

levelplot(as.matrix(test5))

ggplot(test3[test3$Species %in% c("Plains zebra","Grevys zebra"),], aes(x=OTU,y=RRA))+
  geom_col(aes(fill=Species))

```

### 7.4 Community detection

```{r}
test4df = as.data.frame(test5)
test4_bin = test4df %>% mutate_at(vars(MOTU_1:MOTU_94), funs(ifelse(.>0,1,0))) # or motu90
rownames(test4_bin)=rownames(test4)
mod = computeModules(test4_bin, steps=1000)

plotModuleWeb(mod, labsize=0.8, displayBlabels = F)

```


```{r}

#####################################
# project host matrix
g2 = bipartite_projection(g, types=E(g)$type)

hostgraph = g2$proj1
c = igraph::closeness(hostgraph)
b = igraph::betweenness(hostgraph, weights=(1/E(hostgraph)$weight), directed=F)
d = igraph::degree(hostgraph)
e = igraph::eigen_centrality(hostgraph)$vector

netdata = data.frame(species=names(c),c=c, b=b, d=d, e=e)
netlong = pivot_longer(netdata,cols=c(c,b,d,e), names_to="Measure", values_to="Value")

ggplot(netlong, aes(x=reorder(species,-Value), y=Value))+
  geom_col(aes(fill=species), alpha=0.5)+
    theme_bw(base_size=14)+
  theme(axis.text.x = element_text(angle=90, hjust=1,vjust=0))+
  #scale_fill_manual(values=animalcols2)+guides(fill=F)+
  facet_wrap(~Measure, scales="free_y")


plot(hostgraph)
######
```



```{r}

netlong = tidyr::pivot_longer(netdata,cols=c(c,b,d,e), names_to="Measure", values_to="Value")

levels(as.factor(netlong$species))

nodedata = ggplot(netlong, aes(x=reorder(species,-Value), y=Value))+
  geom_col(aes(fill=species), alpha=0.6)+
    theme_bw(base_size=14)+
  theme(axis.text.x = element_text(angle=90, hjust=1,vjust=0))+
  scale_fill_manual(values=animal_colors)+guides(fill=F)+  #[-c(14,18)]
  facet_wrap(~Measure, scales="free_y")
nodedata
#ggsave("g98_plots/nodeplots.png", nodedata, device="png")

phylodists=cophenetic(tree) # patristic distance
# calculate total distance from all others
totaldists = as.data.frame(rowSums(phylodists)/18) # this is Average pairwise distance
totaldists$sciname = rownames(totaldists)

# correct sci name
netdata$sciname = c("Syncerus_caffer",
                    "Madoqua_kirkii",
                    "Giraffa_camelopardalis",
                    "Gazella_granti", 
                    "Alcelaphus_buselaphus",
                    "Aepyceros_melampus",
                    "Oryx_gazella", 
                    "Kobus_ellipsiprymnus", 
                    "Phacochoerus_africanus" ,
                    "Bos_taurus",
                    "Camelus_dromedarius",
                    "Taurotragus_oryx", 
                    "Tragelaphus_strepsiceros",
                    "Equus_asinus",
                    "Equus_grevyi",
                    "Hippopotamus_amphibius",
                    "Equus_burchellii",
                    "Loxodonta_africana")

netdata=dplyr::left_join(netdata,totaldists)
names(netdata)[7]="Distance"

ggplot(netdata, aes(x=log(Distance), y=e))+
  geom_point(size=3,aes(col=species))+
  scale_color_manual(values=animal_colors)+ #[-c(14,18)]
  geom_smooth(method="lm")+guides(col=F)+
  ggrepel::geom_text_repel(aes(label=species))+
  theme_bw()

```

```{r}

netlong2 = tidyr::pivot_longer(netdata,
                               cols=c(c,b,d,e),
                               names_to="Measure",
                               values_to="Value")

netdata = dplyr::left_join(netdata,
                         dplyr::select(RRA_present,
                                       Species,
                                       MSW93_Order:Feeding_Type),
                         by=c("species"="Species")) %>% unique()

netlong2 = dplyr::left_join(netlong2,
                            dplyr::select(netdata,
                                          sciname,
                                          MSW93_Order:Feeding_Type))

measuredict = c(b="Betweenness",c="Closeness", d="Degree", e="Eigenvector")

nodedata2 = ggplot(netlong2, aes(x=Distance, y=Value))+
    theme_bw(base_size=14)+
  #geom_smooth(method="lm",se=F, size=1.5, col="gray20")+
    geom_point(aes(col=MSW93_Family, shape=Feeding_Type),
               alpha=0.6,
               size=4)+
  scale_color_manual(values=c("dodgerblue","goldenrod","green3","maroon3","goldenrod1","greenyellow","green4"))+ # use animalcols2 for values if color
  #ggrepel::geom_text_repel(data=netlong2[which(netlong2$species %in% c("Eland", "Cattle")),], aes(label=species))+
  facet_wrap(~Measure,
             scales="free_y",
             labeller = labeller(Measure=measuredict))+
  labs(x="Mean Pairwise Distance")

nodedata2


plot(hostgraph)


V(hostgraph)$color = ifelse(V(hostgraph)$name %in% c("Hippo"),
                            "greenyellow",
                            ifelse(V(hostgraph)$name %in% c("Elephant"),
                                   "green3",
                                   ifelse(V(hostgraph)$name %in% c("Warthog"),
                                          "green4",
                                          ifelse(V(hostgraph)$name %in% c("Giraffe"),
                                                 "goldenrod1",
                                                 ifelse(V(hostgraph)$name %in% c("Grevys zebra","Plains zebra", "Donkey"),
                                                        "maroon3", "dodgerblue")))))
#V(g)$shape = ifelse(V(g)$type, "square", "circle")
E(hostgraph)$color = "lightgray"
V(hostgraph)$frame.color =  "white"
V(hostgraph)$label.dist = 0


# pdf(here("g98_plots/hostgraph.pdf"), width=8, height=8)
# set.seed(0)
# plot(hostgraph,
#      vertex.label.cex = 1.5,
#      vertex.size=10,
#      vertex.label.color = "gray30",
#      edge.width=E(hostgraph)$weight/2,
#      layout=layout_nicely)
# dev.off()

```

```{r}

netdata = RRA_present %>% 
  group_by(Species) %>% 
  summarize(n=n()) %>% 
  ungroup() %>% 
  left_join(netdata, by=c("Species"="species"))


ggplot(netdata, aes(x=n, y=d))+
  geom_point()+
  geom_smooth(method="lm")


RRA2$Species = as.factor(RRA2$Species)

par(mfrow=c(3,6))
for(i in 1:length(levels(RRA2$Species))){
  RRA2 %>% 
  filter(Species==levels(RRA2$Species)[i]) %>% 
  dplyr::select(MOTU_1:MOTU_557) %>% 
  specaccum() %>% 
  plot(main=levels(RRA2$Species)[i])
}
par(mfrow=c(1,1))

# waterbuck is exceptionally low -- exclude

```


```{r}

# create models with host information
### C
library(caper)
netdata = as.data.frame(netdata)
comp.data=comparative.data(tree, netdata[-which(netdata$Species %in% c("Hippo","Camel","Waterbuck")),], names.col="MSW93_Binomial", vcv.dim=3, warn.dropped=TRUE)


# no phylo
c_mod1 = lm(c ~ Diet + Feeding_Type + log(SocialGrpSize) + log(HomeRange_km2) + log(Body_Mass_kg) + log(Distance), data=comp.data$data)
d_c1 = dredge(c_mod1)
d_c1[1:10]

c_mod = pgls(c ~ Diet + Feeding_Type + log(SocialGrpSize) + log(HomeRange_km2) + log(Body_Mass_kg) + log(Distance), data=comp.data)
summary(c_mod)
d_c = dredge(c_mod)
d_c[1:10,] # 

c_mod = pgls(c ~ log(SocialGrpSize), data=comp.data)
summary(c_mod)


### E
e_mod = pgls(e ~ Diet + Feeding_Type + log(SocialGrpSize) + log(HomeRange_km2) + log(Body_Mass_kg) + log(Distance), data=comp.data)
summary(e_mod)
d_e = dredge(e_mod)
d_e[1:5,] # feeding type
e_mod = pgls(e ~ Feeding_Type, data=comp.data)
summary(e_mod)

e_mod1 = lm(e ~ Diet + Feeding_Type + log(SocialGrpSize) + log(HomeRange_km2) + log(Body_Mass_kg) + log(Distance), data=comp.data$data)
d_e1 = dredge(e_mod1)
d_e1[1:10]

e_mod1 = lm(e ~ Feeding_Type, data=comp.data$data)
summary(e_mod1)

### B
b_mod = pgls(b ~ Diet + Feeding_Type + log(SocialGrpSize) + log(HomeRange_km2) + log(Body_Mass_kg) + log(Distance), data=comp.data, lambda=1, kappa=1, delta=1)

d_b = dredge(b_mod)
d_b[1:5,]# sgs
b_mod = pgls(b ~ 1, data=comp.data, lambda=1, kappa=1,delta=1)
summary(b_mod)

b_mod1 = lm(b ~ Diet + Feeding_Type + log(SocialGrpSize) + log(HomeRange_km2) + log(Body_Mass_kg) + log(Distance), data=comp.data$data)
d_b1 = dredge(b_mod1)
d_b1[1:10]



### D
d_mod = pgls(d ~ Diet+Feeding_Type+log(SocialGrpSize)+log(HomeRange_km2)+log(Body_Mass_kg)+log(Distance), data=comp.data, lambda=1,kappa=1,delta=1)
summary(d_mod)
d_d = dredge(d_mod)
d_d[1:5,] # fd
d_mod = pgls(d ~ log(Distance), data=comp.data, lambda=1,kappa=1,delta=1)
summary(d_mod)

d_mod1 = lm(d ~ Diet+Feeding_Type+log(SocialGrpSize)+log(HomeRange_km2)+log(Body_Mass_kg)+log(Distance), data=comp.data$data)
d_d1 = dredge(d_mod1)
d_d1[1:10]



nodedata2 = ggplot(netlong2, aes(x=log(SocialGrpSize), y=Value))+
    theme_bw(base_size=14)+
  geom_smooth(method="lm", se=F, size=1, col="gray20")+
    geom_point(aes(col=MSW93_Family, shape=Feeding_Type), alpha=0.6, size=4)+
  scale_color_manual(values=c("dodgerblue","goldenrod","green3","maroon3","goldenrod1","greenyellow","green4"))+ # use animalcols2 for values if color
  #ggrepel::geom_text_repel(data=netlong2[which(netlong2$species %in% c("Eland", "Cattle")),], aes(label=species))+
  facet_wrap(~Measure, scales="free_y", labeller = labeller(Measure=measuredict))+
  labs(x="log(Social Group Size)")
nodedata2

nod_dist = netlong2[which(netlong2$Measure %in% c("d","e")),]
nod_soc = netlong2[which(netlong2$Measure %in% c("c","b")),]

not_dist_plot = ggplot(nod_dist, aes(x=Distance, y=Value))+
  theme_bw(base_size=14)+
  geom_smooth(method="lm", se=T, size=1, col="gray20")+
  geom_point(aes(col=MSW93_Family, shape=Feeding_Type), alpha=0.6, size=4)+
  scale_color_manual(values=c("dodgerblue","goldenrod","green3","maroon3","goldenrod1","greenyellow","green4"))+
  facet_wrap(~ Measure, scales="free_y", labeller=labeller(Measure=measuredict))+
  labs(x="Average Pairwise Distance")
not_dist_plot

not_soc_plot = ggplot(nod_soc, aes(x=log(SocialGrpSize), y=Value))+
  theme_bw(base_size=14)+
  geom_smooth(aes(linetype=Measure), method="lm", se=T, size=1, col="gray20")+
  #geom_smooth(data= nod_soc[-which(nod_soc$species%in%c("Warthog","Elephant")),], aes(x=log(SocialGrpSize), y=Value), linetype="dotted", method="lm", color="red", size=2)+
  geom_point(aes(col=MSW93_Family, shape=Feeding_Type), alpha=0.6, size=4)+
  scale_color_manual(values=c("dodgerblue","goldenrod","green3","maroon3","goldenrod1","greenyellow","green4"))+
  facet_wrap(~ Measure, scales="free_y", labeller=labeller(Measure=measuredict))+
  labs(x="log(Social Group Size)")+
  scale_linetype_manual(values=c("dotted", "solid"))+
  guides(linetype=F)
not_soc_plot

nodplots = gridExtra::grid.arrange(not_dist_plot, not_soc_plot, ncol=1)

#ggsave(here("g98_plots/nodplots_2021b.png"), nodplots)



# closeness not defined for disconnected graphs -- use only largest component:
components = igraph::components(hostgraph)
subhost = induced_subgraph(hostgraph, which(components$membership == 1))
plot(subhost)

c2 = igraph::closeness(subhost)
c_data=data.frame(c = c2, species=names(c2))
c_data = c_data[-which(c_data$species %in% c("Camel", "Hippo")),]

nd2 = left_join(c_data, netdata[,c(1,6:18)])
comp.data2 = comparative.data(tree, nd2, names.col="MSW93_Binomial", vcv.dim=3, warn.dropped=TRUE)

c_mod2 = pgls(c ~ Diet+Feeding_Type+log(SocialGrpSize)+log(HomeRange_km2)+log(Body_Mass_kg)+log(Distance), data=comp.data2)
summary(c_mod2)
d_c = dredge(c_mod2)
d_c[1:5,] #

c_mod2 = pgls(c ~ log(SocialGrpSize), data=comp.data2)
summary(c_mod2)


```


Which species important?

```{r}

netdata$sc = BBmisc::normalize(netdata$c, method = "range", range=c(0,1))
netdata$sd = BBmisc::normalize(netdata$d, method="range", range=c(0,1))
netdata$se = BBmisc::normalize(netdata$e, method="range", range=c(0,1))
netdata$sb = BBmisc::normalize(netdata$b, method="range", range=c(0,1))
netdata$tot = netdata$sc+netdata$sd+netdata$se+netdata$sb
neworder = netdata$species[order((netdata$sc)+(netdata$sd)+(netdata$se)+(netdata$sb))]
neworder = netdata$species[order(netdata$tot)]

netlongsc = tidyr::pivot_longer(netdata, cols=c(sc,sb,sd,se), names_to="Measure", values_to="Value")

netlongsc$species = factor(netlongsc$species, levels = neworder)

netlongsc$Measure=as.factor(netlongsc$Measure)
levels(netlongsc$Measure)=c("Betweenness","Closeness","Degree","Eigenvector")
centscore = ggplot(netlongsc, aes(x=species, y=Measure))+
  geom_tile(aes(fill=Value))+
  scale_fill_viridis_c(option="A")+
  labs(x="")+
  theme_bw(base_size=18)+
  theme(aspect.ratio = 1/3.5, axis.text.x=element_text(angle=90,hjust=1, vjust=0))


#ggsave(here("g98_plots/centrality_scores.png"), centscore)
```



### 8. Species Extinction implications

```{r}
# use full dataset
test6=as.matrix(test5)
#test6=as.matrix(samp_mat_2)

net2=as.network(test6, directed=F, loops=F, matrix.type="bipartite")
net2

history=Mostconnected(Network = net2)
history
historydf=history
rownames(test6) # doublecheck and run the appropriate one

#spindex=data.frame(species=rownames(test6), Spp=seq(1,18), IUCN= c("LC(-)","VU(+)","EN(.)", "LC(-)","NT(-)","LC(.)", "D", "D","LC(.)","LC(.)","LC(.)","LC(-)","VU(.)","LC(-)" ,"D", "VU(-)", "LC(.)", "NT(-)"))

# This one 
#spindex=data.frame(species=rownames(test6), Spp=seq(1,18), IUCN= c("LC(-)","EN(.)", "VU(+)", "NT(-)","D","LC(.)", "LC(-)",  "D","LC(.)", "LC(.)","VU(.)","LC(-)", "LC(.)","D", "VU(-)",  "LC(.)","NT(-)", "LC(-)"))


# smaller dataset:
spindex=data.frame(species=rownames(test6), Spp=seq(1,18), IUCN= c("NT(-)", "D","D", "LC(.)", "D", "LC(.)", "VU(+)", "VU(-)", "LC(-)", "EN(.)", "LC(-)","VU(.)", "LC(.)","LC(.)", "LC(.)", "NT(-)", "LC(-)", "LC(-)"))



historydf=(left_join(history,spindex))

historydf$IUCN=factor(historydf$IUCN, levels=c("D","LC(.)","LC(-)","NT(-)","VU(+)","VU(.)","VU(-)","EN(.)"))

exts = ggplot(historydf, aes(x=reorder(species,-Secondary_extinctions), y=Secondary_extinctions))+
  geom_col(aes(fill=IUCN))+
  scale_fill_manual(values=c("#00B2EE", "#ADD8E6", "#FFD39B", "#FFA07A", "#FF7F50", "#EE6363", "#CD5555", "#CD3700"))+
  theme_bw(base_size=18)+
  theme(axis.text.x =element_text(angle = 90, hjust=1, vjust=0))+
  labs(x="Host Species",y="Parasite mOTUs Lost")
exts
#ggsave(here("g98_plots/sp_exts_2021.png"), exts, device="png")

# total extinctions of VU(+) or more:
historydf[which(historydf$IUCN %in% c("VU(.)","VU(+)", "VU(-)", "EN")),]$Secondary_extinctions %>%sum()
25/92
20/92

# nullhyp=RandomExtinctions(Network=net2, nsim=50) # takes a couple of minutes
# 
# # order by IUCN endangerment category
# # EN(s) = grevys, VU(dec)=Giraffe, VU(s)=hippo, VU(inc)=Elephant, NT(dec)=plains,buff, LC(dec)=hartebeest, grants, waterbuck,warthog, LC(stable)=dikdik, impala, oryx, eland, kudu, LC(inc)=donkey, camelhi
# newHist = ExtinctionOrder(Network=net2, Order=c(5,11,17,2,3,9,1,8,10,18,4,6,7,13,15,12,16))
# newHist = ExtinctionOrder(Network=net2, Order=c(2,11,17,9,16,5,3,13,14,12,15,1,10,8,6,18,4))
# ExtinctionPlot(History = history, Variable = "AccSecondaryExtinction");newHist$Graph
# 
# CompareExtinctions(Nullmodel = nullhyp, Hypothesis = newHist)
# # ugh the null includes if parasites go extinct independently
# 
# degree_distribution(net2, name="Test") # exponential -- less vulnerable to removal of most connected nodes, power is more vulnerable

```



### Check for host/parasite co-evolution

```{r}
library(paco)
# prune host tree
tree = drop.tip(tree, "Tragelaphus_scriptus")
treeNJ_MOTU  = treeNJ
treeNJ_MOTU$tip.label = ttlab$MOTU.1

#write.tree(treeNJ_MOTU, "PACO/MOTU_tree")
#write.tree(tree, "PACO/HOST_tree")
#write.csv(gllink2, "PACO/RRAmatrix.csv", row.names=T)

htree = cophenetic(tree)
# need to rename treeNJ tips
ptree = cophenetic(treeNJ_MOTU)

# This is no longer working GRRRRR....

avgRRA$Species
gllink = left_join(avgRRA, dplyr::select(RRA2, Species, MSW93_Binomial))%>%unique()
#gllink2 = as.data.frame(gllink[-which(is.na(gllink$MSW93_Binomial)==T),])
rownames(gllink)=gllink$MSW93_Binomial
# need to specify as a df
gllink2 = as.data.frame(gllink[,-c(1,166)])
rownames(gllink2) = rownames(gllink)

D = prepare_paco_data(H=htree, P=ptree, HP=gllink2)
#D2 = prepare_paco_data(H=htree, P=ptree, HP=gllink2)

D = add_pcoord(D, correction="cailliez")
#D2 = add_pcoord(D2, correction="cailliez")


D = PACo(D, nperm=1000, seed=12, method="r0", symmetric=F)
#D2 = PACo(D2, nperm=1000, seed=12, method="r0", symmetric=F)

D = paco_links(D)
#D2 = paco_links(D2)

res = residuals_paco(D$proc)
plot(res~D$jackknife)

sum(res^2)
print(D$gof)


Dres = data.frame(res=res, connection=names(D$jackknife))
Dres =Dres[order(Dres$res),]
Dres$sp = rep("",dim(Dres)[1])
Dres$MOTU = rep("",dim(Dres)[1])
for(i in 1:length(Dres$res)){
  Dres$sp[i]=strsplit(Dres$connection, split="-")[[i]][1]
  Dres$MOTU[i]=strsplit(Dres$connection, split="-")[[i]][2]
}

ggplot(Dres, aes(x=reorder(sp, -res), y=res))+
  geom_point()+
  theme(axis.text.x = element_text(angle=90, hjust=1, vjust=0))

```


## Tangleplot
```{r}
assoc2 = left_join(mat_1_long, dplyr::select(RRA2, Species, MSW93_Binomial))%>%unique()
assoc2 = assoc2[,c(4,2)]
assoc2$OTU = as.character(assoc2$OTU)
assoc2$MSW93_Binomial = as.character(assoc2$MSW93_Binomial)


assoc2 = assoc2[order(assoc2$MSW93_Binomial),] # That didn't work
assoc3 = assoc2

# update order
species_cols = data.frame(sp = levels(as.factor(assoc2$MSW93_Binomial)), color = animal_colors[c(13,11,3,2,5,16,10, 9,8,12,18,7,4,15,17,1,6,14)])
animal_colors[c()]
assoc3 = left_join(assoc2,species_cols, by=c("MSW93_Binomial"="sp"))

cophyloplot(tree, treeNJ_MOTU, assoc=as.matrix(assoc2), use.edge.length = F, space=100, rotate=F, col=assoc3$color, gap=0.05, length.line=1, font=1, show.tip.label = F)


```


